From 62d57d3a4f8ff92caa910a2f5df703da5b3c73bd Mon Sep 17 00:00:00 2001
From: Keerthana K <keerthanak@vmware.com>
Date: Tue, 5 Jan 2021 10:45:39 +0000
Subject: [PATCH] FIPS crypto self tests

- crypto_self_test module invokes test of crypto algorithms
- Added testmgr test vectors for all the self test combinations
- maintaining a single patch for crypto_self_test module and addition of missing testmgr vectors
- extended test vector coverage for algorithms in LKCM 5.0
- Added cts to crypto self-tests
- Removed rsa(pkcs1pad, sha512), cbc, and ctr from crypto self-tests
- Added ECC pubkey generation and verification success messages

Alexey: add FIPS_NOT_ALLOWED macro
Signed-off-by: Srish Srinivasan <ssrish@vmware.com>

[Srinidhi Rao<srinidhi.rao@broadcom.com> Ported this patch to v6.12.y]
Signed-off-by: srinidhira0 <srinidhi.rao@broadcom.com>

[Shivani Agarwal <shivani.agarwal@broadcom.com> Enable hmac(sha384) self-test]
Signed-off-by: Shivani Agarwal <shivani.agarwal@broadcom.com>
---
 crypto/Kconfig            |   21 +
 crypto/Makefile           |    2 +
 crypto/crypto_self_test.c |  123 ++
 crypto/testmgr.c          |   69 +-
 crypto/testmgr.h          | 2315 ++++++++++++++++++++++++++++++++++++-
 5 files changed, 2517 insertions(+), 13 deletions(-)
 create mode 100644 crypto/crypto_self_test.c

diff --git a/crypto/Kconfig b/crypto/Kconfig
index a779cab66..ff8b81340 100644
--- a/crypto/Kconfig
+++ b/crypto/Kconfig
@@ -235,6 +235,27 @@ config CRYPTO_TEST
 	help
 	  Quick & dirty crypto test module.
 
+config CRYPTO_SELF_TEST
+	bool "Crypto self test"
+	select CRYPTO_MANAGER
+	select CRYPTO_FIPS
+	select CRYPTO_AES
+	select CRYPTO_ECB
+	select CRYPTO_XTS
+	select CRYPTO_GCM
+	select CRYPTO_CCM
+	select CRYPTO_CFB
+	select CRYPTO_CMAC
+	select CRYPTO_CTS
+	select CRYPTO_SHA1
+	select CRYPTO_SHA256
+	select CRYPTO_SHA512
+	select CRYPTO_SHA3
+	select CRYPTO_HMAC
+	select CRYPTO_RSA
+	help
+	  Crypto self test module for FIPS.
+
 config CRYPTO_SIMD
 	tristate
 	select CRYPTO_CRYPTD
diff --git a/crypto/Makefile b/crypto/Makefile
index 4c99e5d37..3b1c1218f 100644
--- a/crypto/Makefile
+++ b/crypto/Makefile
@@ -50,6 +50,8 @@ rsa_generic-y += rsa_helper.o
 rsa_generic-y += rsa-pkcs1pad.o
 obj-$(CONFIG_CRYPTO_RSA) += rsa_generic.o
 
+obj-$(CONFIG_CRYPTO_SELF_TEST) += crypto_self_test.o
+
 $(obj)/ecdsasignature.asn1.o: $(obj)/ecdsasignature.asn1.c $(obj)/ecdsasignature.asn1.h
 $(obj)/ecdsa.o: $(obj)/ecdsasignature.asn1.h
 ecdsa_generic-y += ecdsa.o
diff --git a/crypto/crypto_self_test.c b/crypto/crypto_self_test.c
new file mode 100644
index 000000000..efa26cccd
--- /dev/null
+++ b/crypto/crypto_self_test.c
@@ -0,0 +1,123 @@
+/*
+ * Self test for Crypto Algorithms
+ *
+ * Copyright (c) 2020 VMware
+ *
+ */
+
+#include <linux/module.h>
+#include <linux/fips.h>
+#include <linux/init.h>
+#include <linux/err.h>
+#include <linux/crypto.h>
+MODULE_LICENSE("GPL");
+
+extern int alg_test(const char *driver, const char *alg, u32 type, u32 mask);
+
+struct alg_self_test {
+	char *driver_name;
+	char *alg_name;
+};
+
+static struct alg_self_test __initdata alg_self_test_tbl[]  =
+{
+	{
+		.driver_name = "cts(cbc(aes))",
+		.alg_name = "cts(cbc(aes))",
+	},
+	{
+		.driver_name = "gcm(aes)",
+		.alg_name = "gcm(aes)",
+	},
+	{
+		.driver_name = "ccm(aes)",
+		.alg_name = "ccm(aes)",
+	},
+	{
+		.driver_name = "cfb(aes)",
+		.alg_name = "cfb(aes)",
+	},
+	{
+		.driver_name = "cmac(aes)",
+		.alg_name = "cmac(aes)",
+	},
+	{
+		.driver_name = "ecb(aes)",
+		.alg_name = "ecb(aes)",
+	},
+	{
+		.driver_name = "hmac(sha224)",
+		.alg_name = "hmac(sha224)",
+	},
+	{
+		.driver_name = "hmac(sha384)",
+		.alg_name = "hmac(sha384)",
+	},
+	{
+		.driver_name = "hmac(sha3-224)",
+		.alg_name = "hmac(sha3-224)",
+	},
+	{
+		.driver_name = "hmac(sha3-256)",
+		.alg_name = "hmac(sha3-256)",
+	},
+	{
+		.driver_name = "hmac(sha3-384)",
+		.alg_name = "hmac(sha3-384)",
+	},
+	{
+		.driver_name = "hmac(sha3-512)",
+		.alg_name = "hmac(sha3-512)",
+	},
+	{
+		.driver_name = "pkcs1pad(rsa,sha1)",
+		.alg_name = "pkcs1pad(rsa,sha1)",
+	},
+	{
+		.driver_name = "pkcs1pad(rsa,sha224)",
+		.alg_name = "pkcs1pad(rsa,sha224)",
+	},
+	{
+		.driver_name = "pkcs1pad(rsa,sha256)",
+		.alg_name = "pkcs1pad(rsa,sha256)",
+	},
+	{
+		.driver_name = "pkcs1pad(rsa,sha384)",
+		.alg_name = "pkcs1pad(rsa,sha384)",
+	},
+	{
+		.driver_name = "xts(aes)",
+		.alg_name = "xts(aes)",
+	},
+};
+
+#ifdef FIPS_NOT_ALLOWED
+extern int FIPS_NOT_ALLOWED(char *);
+#else
+#define FIPS_NOT_ALLOWED(algname) 0
+#endif
+
+static int __init crypto_self_test_init(void)
+{
+	int err = -ENOMEM;
+	int i = 0;
+
+	if (!fips_enabled) {
+		return -EAGAIN;
+	}
+	for (i = 0; i < ARRAY_SIZE(alg_self_test_tbl); i++) {
+		if (FIPS_NOT_ALLOWED(alg_self_test_tbl[i].alg_name))
+			continue;
+		err = alg_test(alg_self_test_tbl[i].driver_name, alg_self_test_tbl[i].alg_name, 0, 0);
+		if (err) {
+			printk(KERN_ERR "crypto_self_test failed for %s %s \n",
+					 alg_self_test_tbl[i].driver_name,
+					 alg_self_test_tbl[i].alg_name);
+			goto error;
+		}
+	}
+	return 0;
+error:
+	return err;
+}
+module_init(crypto_self_test_init);
diff --git a/crypto/testmgr.c b/crypto/testmgr.c
index 2f5f6b52b..f3d7cd846 100644
--- a/crypto/testmgr.c
+++ b/crypto/testmgr.c
@@ -3972,6 +3972,7 @@ static int do_test_kpp(struct crypto_kpp *tfm, const struct kpp_testvec *vec,
 		       alg, err);
 		goto free_output;
 	}
+	printk(KERN_INFO "alg: %s: Party A: generate public key test passed.", alg);
 
 	if (vec->genkey) {
 		/* Save party A's public key */
@@ -3989,6 +3990,7 @@ static int do_test_kpp(struct crypto_kpp *tfm, const struct kpp_testvec *vec,
 			err = -EINVAL;
 			goto free_output;
 		}
+		printk(KERN_INFO "alg: %s: Party A: verify public key test passed.", alg);
 	}
 
 	/* Calculate shared secret key by using counter part (b) public key. */
@@ -4896,7 +4898,10 @@ static const struct alg_test_desc alg_test_descs[] = {
 		 * backend cipher -- covered by drbg_nopr_hmac_sha512 test
 		 */
 		.alg = "drbg_nopr_hmac_sha384",
-		.test = alg_test_null,
+		.test = alg_test_drbg,
+		.suite = {
+			.drbg = __VECS(drbg_nopr_hmac_sha384_tv_template)
+		}
 	}, {
 		.alg = "drbg_nopr_hmac_sha512",
 		.test = alg_test_drbg,
@@ -4914,11 +4919,17 @@ static const struct alg_test_desc alg_test_descs[] = {
 	}, {
 		/* covered by drbg_nopr_sha256 test */
 		.alg = "drbg_nopr_sha384",
-		.test = alg_test_null,
+		.test = alg_test_drbg,
+		.suite = {
+			.drbg = __VECS(drbg_nopr_sha384_tv_template)
+		}
 	}, {
 		.alg = "drbg_nopr_sha512",
 		.fips_allowed = 1,
-		.test = alg_test_null,
+		.test = alg_test_drbg,
+		.suite = {
+			.drbg = __VECS(drbg_nopr_sha512_tv_template)
+		}
 	}, {
 		.alg = "drbg_pr_ctr_aes128",
 		.test = alg_test_drbg,
@@ -4930,11 +4941,17 @@ static const struct alg_test_desc alg_test_descs[] = {
 		/* covered by drbg_pr_ctr_aes128 test */
 		.alg = "drbg_pr_ctr_aes192",
 		.fips_allowed = 1,
-		.test = alg_test_null,
+		.test = alg_test_drbg,
+		.suite = {
+			.drbg = __VECS(drbg_pr_ctr_aes192_tv_template)
+		}
 	}, {
 		.alg = "drbg_pr_ctr_aes256",
 		.fips_allowed = 1,
-		.test = alg_test_null,
+		.test = alg_test_drbg,
+		.suite = {
+			.drbg = __VECS(drbg_pr_ctr_aes256_tv_template)
+		}
 	}, {
 		.alg = "drbg_pr_hmac_sha256",
 		.test = alg_test_drbg,
@@ -4945,11 +4962,17 @@ static const struct alg_test_desc alg_test_descs[] = {
 	}, {
 		/* covered by drbg_pr_hmac_sha256 test */
 		.alg = "drbg_pr_hmac_sha384",
-		.test = alg_test_null,
+		.test = alg_test_drbg,
+		.suite = {
+			.drbg = __VECS(drbg_pr_hmac_sha384_tv_template)
+		}
 	}, {
 		.alg = "drbg_pr_hmac_sha512",
-		.test = alg_test_null,
+		.test = alg_test_drbg,
 		.fips_allowed = 1,
+		.suite = {
+			.drbg = __VECS(drbg_pr_hmac_sha512_tv_template)
+		}
 	}, {
 		.alg = "drbg_pr_sha256",
 		.test = alg_test_drbg,
@@ -4960,11 +4983,17 @@ static const struct alg_test_desc alg_test_descs[] = {
 	}, {
 		/* covered by drbg_pr_sha256 test */
 		.alg = "drbg_pr_sha384",
-		.test = alg_test_null,
+		.test = alg_test_drbg,
+		.suite = {
+			.drbg = __VECS(drbg_pr_sha384_tv_template)
+		}
 	}, {
 		.alg = "drbg_pr_sha512",
 		.fips_allowed = 1,
-		.test = alg_test_null,
+		.test = alg_test_drbg,
+		.suite = {
+			.drbg = __VECS(drbg_pr_sha512_tv_template)
+		}
 	}, {
 		.alg = "ecb(aes)",
 		.test = alg_test_skcipher,
@@ -5453,10 +5482,20 @@ static const struct alg_test_desc alg_test_descs[] = {
 		.suite = {
 			.cipher = __VECS(fcrypt_pcbc_tv_template)
 		}
+	}, {
+		.alg = "pkcs1pad(rsa,sha1)",
+		.test = alg_test_akcipher,
+		.fips_allowed = 1,
+		.suite = {
+			.akcipher = __VECS(pkcs1pad_rsa_sha1_tv_template)
+		}
 	}, {
 		.alg = "pkcs1pad(rsa,sha224)",
-		.test = alg_test_null,
+		.test = alg_test_akcipher,
 		.fips_allowed = 1,
+		.suite = {
+			.akcipher = __VECS(pkcs1pad_rsa_sha224_tv_template)
+		}
 	}, {
 		.alg = "pkcs1pad(rsa,sha256)",
 		.test = alg_test_akcipher,
@@ -5478,12 +5517,18 @@ static const struct alg_test_desc alg_test_descs[] = {
 		.fips_allowed = 1,
 	}, {
 		.alg = "pkcs1pad(rsa,sha384)",
-		.test = alg_test_null,
+		.test = alg_test_akcipher,
 		.fips_allowed = 1,
+		.suite = {
+			.akcipher = __VECS(pkcs1pad_rsa_sha384_tv_template)
+		}
 	}, {
 		.alg = "pkcs1pad(rsa,sha512)",
-		.test = alg_test_null,
+		.test = alg_test_akcipher,
 		.fips_allowed = 1,
+		.suite = {
+			.akcipher = __VECS(pkcs1pad_rsa_sha512_tv_template)
+		}
 	}, {
 		.alg = "poly1305",
 		.test = alg_test_hash,
diff --git a/crypto/testmgr.h b/crypto/testmgr.h
index 9b38501a1..347bc0c41 100644
--- a/crypto/testmgr.h
+++ b/crypto/testmgr.h
@@ -1374,6 +1374,726 @@ static const struct akcipher_testvec ecrdsa_tv_template[] = {
 /*
  * PKCS#1 RSA test vectors. Obtained from CAVS testing.
  */
+static const struct akcipher_testvec pkcs1pad_rsa_sha1_tv_template[] = {
+	{
+	.key =
+	"\x30\x82\x0b\x28\x02\x01\x00\x02"
+	"\x82\x02\x01\x00\xb0\xf8\x80\x65"
+	"\x86\xb7\x22\xf4\x14\x07\x04\xf0"
+	"\x00\xf5\x1f\xc9\x03\x28\xad\x8e"
+	"\xc8\xe4\x0d\xeb\x69\x14\xf9\xcc"
+	"\x4a\xfe\x29\x63\x14\xb9\xcc\x6b"
+	"\xbc\x16\x5b\xf4\x27\x94\x4c\xa0"
+	"\xd6\x68\x6b\xfd\xab\x52\xfe\xd2"
+	"\x53\x3a\x3a\x86\xcd\x6a\x58\x4c"
+	"\xd8\x7a\xa3\xaa\x0f\xae\x92\xab"
+	"\x30\x8a\xc6\x96\x7b\xba\x04\xd4"
+	"\xc2\xea\xb4\xca\xfe\x7e\x84\x9a"
+	"\x69\x2e\xef\x11\x34\xd1\x0d\xa2"
+	"\x1b\x5b\x8f\xb2\x94\x54\xbb\xfa"
+	"\xd6\xff\xb0\x98\xec\xf1\x6c\xd1"
+	"\x61\x52\x1f\xcc\x92\x66\x7f\x95"
+	"\x0c\xd4\xe3\x4b\x54\x24\xb0\xc8"
+	"\xa2\x05\x00\xa9\x93\x88\x64\x82"
+	"\xa9\x35\xa8\xc6\x06\xbd\xc9\x10"
+	"\x88\x91\xf9\xe8\x60\x9a\xbe\x17"
+	"\xc3\x9e\xf4\x6b\x92\xf5\xb4\xe7"
+	"\xa3\x24\xcd\xa9\x42\xb8\xfd\x0e"
+	"\x73\x4b\x2a\x2b\xfc\x2d\x75\x48"
+	"\xf8\x8f\x1b\xad\xd8\xab\xa9\x5a"
+	"\x8c\xde\x63\x12\xe8\x89\xb4\x38"
+	"\x5e\xaa\xfe\xf1\x39\xae\xd6\xdf"
+	"\x76\x4f\xb6\xe3\x4a\x5b\xd3\xe1"
+	"\x4d\xe8\x2e\xe5\x9f\xd2\x56\x51"
+	"\x1b\x5f\x66\x15\x4e\x0d\xfa\xed"
+	"\xb5\x74\xdd\x40\x51\xdc\x72\x56"
+	"\x62\xd9\x21\xd2\x73\xc0\x34\x6d"
+	"\x67\x34\xb4\xcb\xfc\xc3\x83\x44"
+	"\xec\xba\x5c\x44\x66\xb5\xb7\xee"
+	"\x50\x3f\xa6\x79\x5c\xb5\x7d\x93"
+	"\xc1\x6c\x05\x34\xef\xb9\xcd\xf6"
+	"\xb1\x8d\x26\xec\x39\x12\x31\x5d"
+	"\x12\x68\xd2\x9c\xe2\x74\xc4\xe5"
+	"\xbf\x46\x6e\xdd\xb6\x92\xee\xd3"
+	"\x9f\x7b\x4b\xe4\x8b\xde\xf2\x3f"
+	"\xd4\x0e\x5c\x11\x40\x9a\x50\xac"
+	"\x81\xab\x9f\x60\xf5\xad\xa7\xa2"
+	"\xb9\xf9\x57\xe8\xc0\xf3\x24\x1f"
+	"\xa1\x89\x95\x31\x19\x32\x66\xd5"
+	"\x8a\xf5\xe4\x66\x65\xf8\xa1\x5a"
+	"\xfb\xb0\xc6\xef\x47\x87\x4a\x60"
+	"\xdd\x0c\xa5\xfe\xd5\x63\xc4\x39"
+	"\x2c\x2e\xd8\xb1\x7b\x53\xa6\x28"
+	"\xa1\x9f\x1b\xf1\xda\x28\xd9\x7e"
+	"\xdf\x15\xc2\x64\x61\x81\x55\x0a"
+	"\x79\x3d\xe6\x9f\x1e\x59\xc7\x74"
+	"\x16\xb0\x75\x48\x02\xb5\x76\x08"
+	"\x64\x10\x53\x6c\xba\xfa\x8e\x93"
+	"\x1b\x55\xb2\x40\x93\x4d\x30\x1b"
+	"\x1f\x11\xc0\x8a\x7e\x73\xe1\x45"
+	"\x46\x95\x5f\x2a\xde\x09\xe5\x11"
+	"\x1f\xa0\xb3\xa2\x29\x08\xac\x96"
+	"\x28\x3b\x8a\xa1\x1e\xb2\x22\x8c"
+	"\x5e\xc1\xe5\x26\x5b\x9f\x62\xa1"
+	"\xb2\x01\xa2\xba\xa0\x75\xfa\xd3"
+	"\xa4\x23\xd4\x26\x82\xeb\x39\x0a"
+	"\x0c\xc8\x0d\x8a\xc3\xda\x92\x62"
+	"\xd1\x5a\xe1\x3f\x4d\xd7\xf2\x70"
+	"\x32\x9f\x4e\xc9\x24\x18\x34\xfc"
+	"\x5e\xc6\x05\x00\xa8\x3d\x86\x02"
+	"\xf8\x10\xa4\x4b\x6b\xb9\x2c\x84"
+	"\x49\xec\x22\x45\x02\x82\x02\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x01\x00\x01"
+	"\x02\x82\x02\x00\x2a\x65\x1d\xbd"
+	"\x34\xac\xa2\xbe\xfe\x1d\x40\x5d"
+	"\x8e\x9d\x2a\x64\x29\x51\x91\x62"
+	"\x67\x55\xda\xcf\xce\x78\xab\xee"
+	"\xf8\xf8\x32\x49\xfe\x29\x14\x83"
+	"\xaa\xa7\x60\x97\xf8\xf3\x59\xf8"
+	"\x1d\xe5\x1e\x19\xc2\xf5\x3a\xfe"
+	"\x7f\xd9\xaf\x33\xa3\xfa\x19\xf5"
+	"\x2e\xd7\x12\x66\x27\x23\x97\x17"
+	"\x53\xa5\xab\x35\x59\xa6\x92\x89"
+	"\xb5\xb8\x2c\x5f\x84\xe0\x22\x3c"
+	"\x2b\x7e\x86\xdf\x04\x5c\x20\x7b"
+	"\xf9\x39\xd2\x0e\xe1\xe1\xfa\xb6"
+	"\x78\xd3\x00\x27\x62\x7b\xe3\xa9"
+	"\xb5\x3a\xcb\x4b\x62\xb7\x13\x51"
+	"\x4d\xa5\xa2\x5d\xe5\x66\x74\xc3"
+	"\x5b\xac\xff\x1f\xa0\xac\x48\x07"
+	"\xd5\x29\x79\xa9\x16\xec\x58\x42"
+	"\x91\xb5\x06\xbd\x58\x87\x39\x8a"
+	"\xcf\xde\x40\x87\x94\xc9\x2f\xf1"
+	"\x5d\xef\xef\x24\xe1\x4b\x17\xb4"
+	"\x62\x82\x5b\x33\xf6\x51\x9d\x8b"
+	"\x41\xa0\x66\xed\x1a\xbe\xe4\xef"
+	"\x40\x57\xf6\x3c\x0d\x61\x0f\xbf"
+	"\x81\x4e\x4d\x8c\xd4\x58\xc2\x38"
+	"\xfe\x4a\xd2\x71\x58\xda\xee\xc3"
+	"\x46\xba\x9c\xa1\x7d\xcc\x11\xc9"
+	"\x94\x59\x22\x47\x70\x4f\x81\x11"
+	"\x79\x2a\xe9\xe6\x02\xb6\x60\xdd"
+	"\xb9\x08\x73\x45\xcb\x5c\xfc\x32"
+	"\x49\x4b\x23\x14\xcb\x14\xbb\x59"
+	"\xc3\xac\x51\x76\x41\x74\xbe\xff"
+	"\x64\x52\x1b\x3b\x09\x76\xa1\xaa"
+	"\xbb\x1f\xda\xe4\xb9\x8d\xd7\x7d"
+	"\x6b\xad\x3f\x87\xac\xe5\x5f\x6d"
+	"\x86\x66\xf1\x1a\x48\xf0\xba\xe8"
+	"\x9c\x49\xb7\x31\x1b\x07\xad\x28"
+	"\x41\x0a\x77\x75\x91\x13\x26\x82"
+	"\x7f\x4a\x9f\x62\x96\x9f\x80\x54"
+	"\xa4\xa5\x33\xd0\x6d\x84\x3c\x3f"
+	"\x9e\xd8\x65\x98\x63\xe2\x77\x76"
+	"\x8f\x89\x60\x68\xcf\x04\x63\x35"
+	"\xbd\xb8\xc0\x5b\xe8\x2e\xb6\xb0"
+	"\x5a\x06\x6c\x53\x02\x98\x1c\x51"
+	"\x7f\x6b\x85\x46\x66\x7a\x08\xc9"
+	"\x67\xca\x30\x40\xca\x5c\x4b\x85"
+	"\x68\x4b\x7b\x9d\x29\x50\x5b\x4e"
+	"\xac\xcb\x9b\xa4\xb9\xe2\x47\x29"
+	"\xdb\x08\x92\x20\xba\x9d\xf8\x11"
+	"\x95\x6c\xc6\x22\x57\xb8\xd6\x76"
+	"\x5e\x64\x78\x2e\x62\x46\x8c\xd8"
+	"\x41\xbc\x26\x2a\x84\xa4\x7b\x62"
+	"\x23\x2e\x78\x12\xb5\x7b\x9b\x89"
+	"\x1b\xae\x7a\x45\x4b\x3c\xf4\x15"
+	"\xac\xa4\xe8\xe3\xe0\x64\x80\x1a"
+	"\x31\x95\xaf\xc5\xd9\x55\x4d\x4e"
+	"\x68\x47\x0d\xa7\x5e\xdd\x5d\x70"
+	"\xe9\xc2\x0a\x32\x7a\x11\x62\x92"
+	"\x5e\x37\x06\x67\x4d\x68\x87\x65"
+	"\xaa\xdb\x41\x35\x6e\x36\xba\xb6"
+	"\x64\x7c\x79\x84\xbe\x35\x16\x43"
+	"\xdf\xf9\xd9\x16\xc1\x49\x62\x49"
+	"\x5f\x14\x0a\x90\x5c\x60\xaf\xee"
+	"\x09\xea\xce\x87\x71\x11\x24\xce"
+	"\xf3\x29\x61\x81\x02\x82\x01\x01"
+	"\x00\xe3\xe0\xd4\x64\xb3\xa8\x8f"
+	"\x16\xd0\x33\xfb\xfc\xca\xc5\x4c"
+	"\xbc\xb7\xc6\xe8\x6c\x71\x55\xbf"
+	"\xe1\xfa\x6e\x12\x48\x3d\x94\xae"
+	"\xb9\x98\xf4\xe1\x59\x24\xfd\x79"
+	"\x73\x6e\xc8\xad\x90\x83\x38\x02"
+	"\x8d\x27\x4c\x87\x56\x5c\xed\x66"
+	"\xbf\xa0\x51\x53\xfa\x7c\x87\x0b"
+	"\xe5\x79\xf5\xda\x3f\x2b\x11\xf5"
+	"\xb4\x1c\xed\x6a\x11\xc6\xfe\x73"
+	"\x2f\x86\x76\x20\x46\x0f\x9b\x84"
+	"\x76\x33\xaa\xc5\xb4\xfa\xe9\x16"
+	"\xbb\x10\xe5\x0c\x3f\x02\x68\x69"
+	"\x10\x81\x30\x4a\x90\x29\xd7\xfb"
+	"\x48\xc7\x98\x59\xb5\xbf\xae\xf6"
+	"\x76\x8f\x6b\x12\xaf\x0d\xc6\x2e"
+	"\x2c\xfe\x35\xdf\x62\x01\xb7\xee"
+	"\x49\xd5\x29\x49\x43\xe4\xfe\x3c"
+	"\x03\xeb\x7a\x04\x6a\xbc\xaf\x44"
+	"\xab\x7d\xe6\x65\xa3\x98\x9c\x77"
+	"\x03\x18\x4e\x90\x13\x3a\x47\x81"
+	"\x8b\x1f\x5d\x11\x6e\xfb\xe2\x56"
+	"\x52\x96\x38\x59\x48\xea\x4e\x9e"
+	"\x25\xe8\xa2\x20\x34\xa3\x90\x1f"
+	"\x01\xee\xa9\x73\xd9\xeb\xca\xbb"
+	"\x36\x76\xc7\x64\xeb\x74\xf7\x10"
+	"\x58\xd1\x77\x41\x91\x1e\xf5\x2b"
+	"\x86\x34\x3f\xe8\x34\x1e\x23\x01"
+	"\xf5\xd6\x36\x40\x60\x50\xec\xd5"
+	"\x22\x57\x75\x77\x75\x3d\xa5\x91"
+	"\xc5\x25\xf0\xc4\xe1\xab\x5b\xa0"
+	"\xb7\x48\xbb\x44\xff\x82\x2c\x40"
+	"\x5d\x02\x82\x01\x01\x00\xc6\xcf"
+	"\x64\x59\xf5\x21\xdf\x0d\x49\xcf"
+	"\x81\x4e\xb4\x73\x45\x65\xa2\x15"
+	"\x89\x33\xd8\x1d\x73\xa1\xa1\xbb"
+	"\x5e\x21\x34\x6e\xb5\xb9\x81\x54"
+	"\x9b\xcb\xc6\x8f\xf2\x23\xc0\x21"
+	"\xf3\x31\xcc\x27\xd5\xae\x43\xec"
+	"\xf2\x8e\x16\xc1\x58\x2b\xe9\x14"
+	"\x6c\x1e\x3e\x3a\x96\xe0\x3a\x7c"
+	"\x39\x19\x07\x00\x97\x8f\x84\xa1"
+	"\x91\x4e\xce\x3b\x2c\xc8\xe5\xfd"
+	"\x45\xad\xcb\x08\xf5\x9d\xf2\x38"
+	"\x80\x14\x5f\xef\xf4\xfb\x2d\xb1"
+	"\x47\x47\xf6\xf9\x1d\x12\xbf\x9d"
+	"\x8e\x0c\xe4\xd0\xe9\x9a\x0e\x5d"
+	"\x8c\xc7\xe1\xba\x89\x22\xa0\x56"
+	"\x8d\x4b\xcb\xb2\x77\xa8\x82\x63"
+	"\x21\x04\x7d\xbe\xe0\x9f\xa3\xbe"
+	"\xaf\xe0\xeb\x6c\xc6\x4a\x71\x1b"
+	"\xfa\xf2\x95\x7e\xf8\x44\x9e\x30"
+	"\xb8\x7c\xe5\x9b\xc7\xe9\x67\x58"
+	"\xe4\x53\xa6\x90\x2a\x65\x26\x68"
+	"\x7b\xee\x4e\xe1\x09\xbf\x58\x3c"
+	"\xc8\x76\xa1\x85\xd9\x71\x9a\xf9"
+	"\xeb\xb7\x0f\xb3\x14\x5e\xca\xac"
+	"\x8d\x4c\xd7\x94\xb2\x17\x1b\x13"
+	"\x04\xb9\x89\x7c\x64\x58\x2a\x1e"
+	"\x40\x28\xfe\x81\x7b\xcd\x64\x4b"
+	"\x24\x96\x6a\xe7\x8c\xd3\x49\xaf"
+	"\x51\xc6\xfd\xb7\x06\x00\x60\x7e"
+	"\x6b\x0a\x88\x06\x87\xa4\x03\x6c"
+	"\x94\x04\x3e\xcd\x02\x3c\xcf\x4d"
+	"\x1e\xa1\x69\xe3\x6b\x09\x02\x82"
+	"\x01\x01\x00\x84\x28\x48\x05\x7f"
+	"\x0c\x2a\xb0\x18\x5f\x8d\xac\xb6"
+	"\x0c\x66\xcc\xab\xe8\x77\xb3\x2c"
+	"\xa7\x86\x79\x10\x3b\xf5\xee\x5f"
+	"\x88\x73\x2c\xc0\x40\x84\xd0\xa6"
+	"\x5b\x3a\xac\x6f\x7b\x82\xed\x64"
+	"\xac\x7f\x0e\x4d\xd4\x5c\x49\x01"
+	"\xeb\xcc\xc3\x25\xbf\x04\xed\x28"
+	"\x7b\x49\xb2\xf4\x60\x59\x0a\x79"
+	"\x78\x18\xff\x34\x27\x9c\x74\x62"
+	"\xeb\xfe\xf1\x79\x11\xd8\xf5\xc8"
+	"\xe7\x70\x92\xfc\xb3\xf5\xa5\x58"
+	"\x3f\x06\x27\xc7\x4c\x8f\x25\x9b"
+	"\xc7\xe1\xe6\xe8\x85\x91\x38\xba"
+	"\x8b\xbf\xf8\x1d\x89\x85\xbd\x9a"
+	"\x45\x66\x3a\xac\x42\x74\xbb\x81"
+	"\x49\x86\x4e\x21\xc0\x2c\xa5\x32"
+	"\x49\xf1\xb0\x70\x27\x59\x8e\x3e"
+	"\xa8\x13\x36\x52\xc2\x98\xfd\x2e"
+	"\x20\x97\xba\xbd\x71\x79\x50\xbf"
+	"\x99\x3e\x8c\x54\xf4\x7b\x2a\x2b"
+	"\x8a\x7c\xde\xaa\x79\xd7\x3a\x7e"
+	"\x6a\x6d\x95\xc7\xf9\x36\x0e\xcb"
+	"\x94\x7c\xb3\x8f\x8f\x37\x0e\x50"
+	"\x2d\xad\x4d\xf9\x93\xdc\xd4\x6d"
+	"\xa3\x4d\x9f\xf0\x93\x03\x3c\x6a"
+	"\xc1\x3e\x81\xc4\x39\x5d\x95\x44"
+	"\x56\xfb\xbf\x01\x5e\x6e\xd4\x8b"
+	"\x05\xd2\x85\x38\x8e\x8e\x17\x4d"
+	"\xd7\x8c\x64\x36\x44\xb1\xdb\x14"
+	"\xbd\x89\x24\xc5\x34\x35\x7a\x8b"
+	"\x5e\xc0\x7e\x8b\x77\xd0\x34\x3e"
+	"\xd2\x5b\x21\x02\x82\x01\x01\x00"
+	"\xc4\x7c\x86\x1e\x8e\x31\xca\xa3"
+	"\xf4\x75\xfb\xc3\x50\x7e\xb8\x70"
+	"\x88\xef\xa5\xe9\x32\x47\x71\x57"
+	"\xa6\x63\x6a\xd5\x80\x5f\x86\x84"
+	"\xef\x78\x32\x7e\x04\x21\x91\xdb"
+	"\x99\x71\x34\x4c\xdc\x0b\xe6\x98"
+	"\xd3\x93\x23\xea\x20\x3d\x38\x2a"
+	"\x35\xb2\xb4\x64\x73\x76\x2e\x55"
+	"\x30\x18\x43\xca\x23\x6c\x87\x5e"
+	"\x15\x64\x25\x1b\x70\xa1\x75\x52"
+	"\x9c\x53\x63\x81\x2f\x30\x0c\xa7"
+	"\xec\x93\xc0\xa9\xe2\x63\x62\x88"
+	"\xb8\x91\x03\xf4\xf0\xf7\x25\xaf"
+	"\xb8\xe7\x3b\x54\x2c\xf4\xf8\x5b"
+	"\x04\xa5\x9a\x94\x95\xd8\x68\xde"
+	"\x6a\xfe\xff\xa8\x90\x8a\x32\xce"
+	"\xbd\x06\xc1\xc4\x11\x8f\x0b\xab"
+	"\x5b\x07\x04\x06\x54\x46\xed\xf6"
+	"\xf6\xb1\x03\xa4\xaa\xf1\x90\xe7"
+	"\x28\x3c\xd8\xa8\xbf\xb8\xe5\xac"
+	"\x1d\x32\xf2\x7e\x66\x19\x07\x7d"
+	"\x6e\xa7\x47\x0c\xb0\xb2\xa2\x92"
+	"\x94\xdf\x85\xf4\x11\x11\x1b\x65"
+	"\xb7\x18\x17\xc9\x95\xdf\x94\x5b"
+	"\x57\x0d\x90\x9d\x08\x57\x72\x67"
+	"\x79\x9f\xa3\x44\xc7\xc7\xc3\xd8"
+	"\x8a\xcf\xd8\xf3\x6b\x0d\x87\xd0"
+	"\xe7\xb5\x88\x54\x8d\x15\x89\xcd"
+	"\xec\x7e\x34\x61\x69\xb1\x70\x62"
+	"\x00\xfc\x10\x48\x99\x45\x72\xdb"
+	"\x3b\x73\x1d\x1a\xd1\x02\x86\x9a"
+	"\x98\xfb\x0c\x31\xd5\x58\x29\x19"
+	"\x02\x82\x01\x00\x00\x87\x87\x9b"
+	"\x3d\x45\x6c\xfb\x63\x20\xa1\xb9"
+	"\xf5\x69\xce\x6b\x96\xae\x69\xf0"
+	"\xe1\xf2\x8b\x39\x2a\x18\xc8\xca"
+	"\x58\xc1\x30\x2f\x15\xca\x93\x89"
+	"\x38\x37\x83\x85\x6e\x09\x5b\xad"
+	"\xac\x43\x23\xa3\x19\xfd\x1d\xc8"
+	"\xa2\x8f\x75\x55\xc8\xf7\x0e\x37"
+	"\xa8\x52\xd1\xb6\x94\x03\xc9\xe1"
+	"\xbf\xde\x4f\xd9\x3c\xa0\xc4\x2d"
+	"\x7d\x6a\x57\x02\xb6\x17\x05\xd7"
+	"\x0d\x43\xce\xf0\x2b\x95\xb3\x31"
+	"\xdf\x96\x03\x36\x5e\x5d\xed\xbd"
+	"\x71\xc3\xf2\x64\x2c\x3d\x3b\x64"
+	"\xfa\x3b\x83\x94\x6d\x07\xda\x52"
+	"\x44\x60\x51\xe5\x4b\xcf\x1d\xec"
+	"\x50\x50\xa6\xea\x39\x40\x18\x17"
+	"\x1d\xd8\xd2\xdf\xc9\x4e\x52\xf4"
+	"\x92\x95\x49\xd6\x3d\x1e\x03\xa2"
+	"\x85\xd7\x5c\xa4\x84\x6c\x80\x22"
+	"\x63\x14\x6c\x41\x20\x23\xbb\x54"
+	"\xc7\x05\x5f\xb7\xb4\x4e\xbd\x79"
+	"\x7d\xeb\xb6\x47\xcc\x70\xeb\x0e"
+	"\x98\x12\x09\x44\x7e\x22\x34\x6a"
+	"\x11\x17\x2f\xdd\xe2\xac\x51\xe9"
+	"\x97\x1c\xb1\xe9\xa2\x3e\xb2\xce"
+	"\x25\x9d\xb6\x67\xed\x11\x93\xa8"
+	"\x62\x49\xcb\x35\x41\xd2\xdd\x9b"
+	"\xb5\xb4\xcb\xcf\xd3\xd5\x09\xc4"
+	"\xfe\xfe\x49\x01\xc7\x0a\xe8\x9d"
+	"\xe3\x09\xa9\x3d\x73\x7f\x20\x8f"
+	"\xbb\x56\xda\x27\x28\x39\xbc\xf6"
+	"\x59\xfe\x77\xb2\x00",
+	.key_len = 2861,
+	/*
+	 * m is SHA1 hash of following message:
+	 * "\x15\x1a\x49\x51\x70\x2b\x73\xd2"
+	 * "\xe6\xd0\xdc\x0d\x75\x72\x3a\xfd"
+	 * "\x3f\x1c\xa8\x78\x3a\x3e\x1e\xfd"
+	 * "\x69\x6e\xf5\x77\x45\x92\xce\x9b"
+	 * "\x4b\xf7\x61\x12\xba\xbc\x18\xc4"
+	 * "\x0d\x53\x5f\x79\x4e\x18\xfb\x42"
+	 * "\x88\x7c\x78\x71\xae\x8a\xae\x76"
+	 * "\x2c\x03\x0d\x89\xef\x80\x25\x46"
+	 * "\x4a\x2a\x26\x70\xb5\xc3\x7a\xb7"
+	 * "\x8e\x57\x03\xe9\xb5\x4a\x01\x76"
+	 * "\xf7\xe5\xf8\x6b\xd8\x18\xc6\x71"
+	 * "\x68\x17\xd5\x34\x73\xa6\x1b\x10"
+	 * "\x3c\x14\xc4\x50\x8f\xf3\x28\x81"
+	 * "\xe8\x51\x8a\x75\xfc\x16\x8d\x01"
+	 * "\x37\x81\x30\x0b\x91\x62\x5b\x0f"
+	 * "\x7a\xb6\x64\xde\x44\x62\x08\x18"
+	 * "\xcc\xff\x00\x42\x4c\x39\xf8\x28"
+	 * "\xf0\x89\xce\x33\x28\x41\x28\xcc"
+	 * "\xf5\x19\x96\x7c\xbf\xe7\x10\x3d"
+	 * "\x57\x10\x89\x88\x1b\x19\x0b\xb6"
+	 * "\x84\xb2\xb6\x24\xe3\x82\x1b\xce"
+	 * "\xfe\xf0\xf3\xa1\x9f\xb2\x1f\x04"
+	 * "\x3f\x7f\xf7\x4f\x19\x8d\xcb\x83"
+	 * "\x24\xe3\xc5\x67\x08\xf6\x18\xe3"
+	 * "\x2b\x6d\xe6\x47\x46\x67\x4c\xec"
+	 * "\x69\xaf\x0a\x8b\x6a\x35\x20\x1d"
+	 * "\x77\x91\x7f\x34\xfd\x36\x24\x6b"
+	 * "\xfa\xfd\x8c\x2f\x6f\x85\x7e\x08"
+	 * "\x7b\x5e\xd6\xc0\xd0\x97\x95\xbe"
+	 * "\xb4\xf4\x86\x3d\xf8\xc6\x80\xb1"
+	 * "\x12\x35\xec\x86\xff\x94\x41\x02"
+	 * "\x2e\xb7\xde\xdc\x25\x00\xe0\xf0"
+	 */
+	.m =
+	"\x11\xf6\xe7\xb6\x78\x4b\xaf\x4e"
+	"\xe2\x48\x2a\x00\x48\x9c\xc9\xf6"
+	"\x4e\xc0\x6f\x13",
+	.m_size = 20,
+	.c =
+	"\x59\x13\xb0\xd9\x68\xe8\xe3\x0b"
+	"\x46\x0f\xdf\x4c\x0d\x23\x83\x86"
+	"\xa0\x16\x46\xc8\x9e\xca\x9a\xda"
+	"\x1e\xc4\x69\xd3\xc5\x84\x3d\x32"
+	"\x73\xf2\x4c\x85\x22\x3f\x07\xd0"
+	"\xfc\x0e\x36\xdb\x3c\xdd\x07\xb5"
+	"\x09\xd0\x0b\x2e\xe1\x04\x0b\x02"
+	"\xb5\x68\x91\xfa\x17\x83\xf7\xb3"
+	"\x40\xea\x86\xc4\x70\x6e\xa0\x1b"
+	"\x5c\x9b\xbe\x8a\xf9\x2d\xb6\x28"
+	"\xeb\xcc\x73\xbb\x60\x50\x51\xfa"
+	"\xbc\x6c\x52\x0c\x1c\x13\x37\xaa"
+	"\xff\x06\xc0\xc8\xff\x75\x1b\x17"
+	"\x01\xe7\x59\x6e\xa6\x71\x29\xbe"
+	"\xf2\x85\xe9\x2e\xb1\x09\x90\xe6"
+	"\xab\x1d\x99\xf2\xce\x66\x6d\xd6"
+	"\xf9\x02\x76\x00\x2f\xd3\xd3\x2e"
+	"\x98\x2d\x2c\x92\x4e\x3f\xc3\xb8"
+	"\x29\x6d\x11\x30\x6f\xa4\x33\x87"
+	"\x6b\xe9\x69\x99\xea\x4d\x99\x60"
+	"\x37\xd9\xbb\x1b\x54\x63\xab\x62"
+	"\x00\xc3\xdd\x2b\x33\xaf\xf2\x22"
+	"\xa9\xd8\x66\xe4\x7b\x2b\xe6\xf7"
+	"\x9e\x4a\x13\xc8\x6b\x3f\x0a\xd9"
+	"\x31\x9c\xb2\x80\xf2\x00\x43\xd3"
+	"\x70\x49\x89\xb4\x09\x6d\xde\xc9"
+	"\x07\x84\x77\x85\xbc\x84\x47\xcf"
+	"\x4f\xa8\x3b\x11\xc0\x41\x33\xbb"
+	"\x54\x11\xa2\x66\xdd\x44\xd0\x87"
+	"\xc5\xea\x6a\xc1\x08\x8d\x9a\x0b"
+	"\x8b\x9e\xd8\x5c\x03\xf6\x13\x4a"
+	"\xd8\xdb\x5b\xcd\x35\x91\xbc\x69"
+	"\x40\x83\x8a\xbe\xba\x45\xd2\xf4"
+	"\x08\x04\xc4\x9f\x2f\xf1\xe2\x0b"
+	"\xb7\xec\x91\x5a\xe4\xbe\xa3\xb6"
+	"\xc9\xec\x79\x3c\x60\x0a\x6a\xe9"
+	"\x93\xa4\xec\x3e\xbb\x84\x59\xad"
+	"\x65\x09\x10\x0b\xcd\xa7\x1b\xdf"
+	"\xec\x15\x2f\xb1\xac\x5c\xf6\xab"
+	"\x44\x99\xa3\xae\x82\xeb\x01\x7a"
+	"\x17\x4a\xa6\xe2\x4f\x4f\x77\x0b"
+	"\xd6\x25\x54\x82\x53\xa4\xf9\x23"
+	"\x41\xa0\xa3\x4a\x88\xc2\xf8\xfa"
+	"\x51\x0a\xd7\x0f\xbd\x1e\xf3\x52"
+	"\x2a\x21\x3d\xb8\x04\xa5\xbc\x5b"
+	"\xa2\x61\xcf\x89\xc0\xce\x9b\x6a"
+	"\xd6\x29\xbe\x82\x7f\x36\xd1\x64"
+	"\x6e\x7c\x35\x0c\x3f\x55\xb1\x23"
+	"\x2f\x84\xbb\x56\x52\xe5\x0b\x0e"
+	"\x88\xac\x17\xb1\x6c\x1c\x60\xfc"
+	"\x82\xae\xe5\x3a\x1d\x3f\x40\x48"
+	"\x5d\xbc\xc0\xba\x89\x9b\x26\x33"
+	"\xec\x18\x3b\xa8\xfc\x5b\x35\xc0"
+	"\x25\xd1\xc8\x79\x72\x74\x8e\x5a"
+	"\xa3\xf1\x72\x63\x72\xec\xfd\xc7"
+	"\xe4\x14\x0d\xeb\xd0\x4b\x17\xd1"
+	"\x8a\x28\x43\x8f\x30\xce\xf6\x87"
+	"\xe3\xd5\xe3\xce\xd0\x54\xaa\x5d"
+	"\x65\x7b\xf7\x67\xeb\x5e\x5d\xf4"
+	"\x73\x2a\x86\x71\x14\x09\x25\x1c"
+	"\x48\xd9\x98\xc8\x0a\x15\x58\x62"
+	"\x51\xe7\x7c\xce\x0a\xf0\x6d\xd9"
+	"\x20\x68\x11\x8c\x3b\x2d\xd9\xcb"
+	"\x13\xe1\x1c\xe3\x2c\x80\xbb\x9a",
+	.c_size = 512,
+	.siggen_sigver_test = true,
+	}
+};
+
+static const struct akcipher_testvec pkcs1pad_rsa_sha224_tv_template[] = {
+	{
+	.key =
+	"\x30\x82\x05\xa2\x02\x01\x00\x02"
+	"\x82\x01\x01\x00\xcf\x6d\xc3\xce"
+	"\x2b\x18\xd8\x74\x0a\x4d\xe6\x4a"
+	"\x49\xfb\x26\x43\xbe\xf1\x07\xc5"
+	"\xdb\x64\xb7\x89\xe3\xcf\xcc\xc9"
+	"\x2b\x7b\xae\xe4\x80\x53\x4e\x5e"
+	"\x07\x52\x27\x52\xab\x8f\x6e\x09"
+	"\x8f\xff\x9f\x89\x34\x59\xd9\x7c"
+	"\xa1\x91\x36\x30\x07\x73\x44\xda"
+	"\xf7\x29\x6d\xa1\x32\x0b\x0c\xdd"
+	"\x27\x11\xb4\x09\xa9\x9c\x8f\x7f"
+	"\x95\xb1\x0c\x16\x03\x68\x0e\x55"
+	"\x2e\xbb\x3d\x0b\x59\xf6\x37\x88"
+	"\xb6\xb7\x13\xb0\x1e\xb1\xe0\xef"
+	"\x04\x32\x2c\xd9\x25\x4b\xd5\x77"
+	"\x10\x3c\xc3\x4a\x53\x72\x96\xc9"
+	"\x2c\x9f\x60\xea\x4c\xf0\x1d\x4b"
+	"\x0b\xf9\xc4\xb4\x90\x8f\xbd\x3e"
+	"\xd4\x10\x87\xba\x7b\x0c\x3f\x6d"
+	"\x84\x7a\xa0\x1e\xb2\x11\x1b\x75"
+	"\x9a\x9b\xf6\x09\xbe\x24\xd0\xee"
+	"\x65\x64\x3f\x73\x18\x77\x52\xf8"
+	"\x8a\x83\xf8\xab\x7c\xff\xe3\x91"
+	"\x8c\x5a\x6a\x1d\x96\xfd\x27\xd7"
+	"\x64\xfa\xe9\x2a\xd7\x36\x9e\x3a"
+	"\x81\xaf\xcf\xc1\x11\x1d\x45\x57"
+	"\xfe\xdf\x4e\x90\x49\x70\x45\x39"
+	"\xab\x9c\x51\x3a\x08\xb9\x3b\x46"
+	"\xe5\x71\xeb\x1c\xd0\x62\x9d\xd4"
+	"\xb7\x4b\xb0\xfc\xaf\x36\xdd\xab"
+	"\x80\xc3\x45\xb0\x09\x2e\xcc\x3c"
+	"\x41\x68\x89\x33\x23\x27\xa7\x2f"
+	"\x49\xd1\x78\x00\xa5\x35\xa9\xce"
+	"\xd8\xde\x66\xa9\x02\x82\x01\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x01\x00\x01"
+	"\x02\x82\x01\x01\x00\xc7\xe2\x1d"
+	"\x2a\x61\xcb\x4a\x61\xdf\xc0\x1b"
+	"\x67\x81\x0c\xfc\x5d\x91\xf5\x10"
+	"\x29\x70\x76\x08\x8b\x5e\xb4\xd2"
+	"\xc1\x14\x1e\xd6\x74\xa8\xd8\x1e"
+	"\x0f\x6b\xfc\x95\x9a\x0b\x6a\x66"
+	"\x7d\xde\x7e\x54\x8c\xd1\x9e\xb7"
+	"\x7e\x6a\xee\xd4\x89\xc9\xd4\x21"
+	"\x08\xe3\xdd\xcf\x1f\xe3\x6a\x54"
+	"\x96\x13\xdc\xf2\x61\xd1\x48\xe5"
+	"\xda\xeb\x33\xb5\xd2\xcc\xab\x8c"
+	"\x0c\x37\xc1\x04\x5d\x18\xee\xb3"
+	"\xab\x11\x6f\x30\x10\xaa\x33\x7c"
+	"\xee\xb6\x5b\x67\x6e\x44\x8b\x64"
+	"\x03\x9c\xe9\x00\xc3\x71\xd0\x5d"
+	"\x4f\x34\xa9\xcc\xa9\x5b\x44\x92"
+	"\xe5\xcb\x7f\x7f\x35\x48\x5c\x72"
+	"\x1e\x24\x62\xa5\x6c\x46\x7a\x11"
+	"\x31\xf3\xbf\xcf\x84\x1b\x79\x55"
+	"\xd0\xa2\xf2\xdf\x99\xe5\x69\xb3"
+	"\x8c\x90\xbd\xa5\x80\xad\x88\x3a"
+	"\xdb\x94\xe9\x1e\x88\xfe\x75\xae"
+	"\xef\x62\xd1\xd4\x91\x0e\x48\x37"
+	"\x06\x51\x89\xcb\x65\x8e\x44\x7d"
+	"\x39\xad\xdd\xc3\x71\x6e\x6a\x63"
+	"\x77\x95\x4e\x77\x3f\xc5\x59\x31"
+	"\x16\x06\xbb\x7f\xff\x53\xeb\x59"
+	"\x97\x19\xaf\x8e\x13\xb2\xe9\xcb"
+	"\x67\x9e\x43\xc1\x5b\x99\x98\x25"
+	"\x56\x6d\x66\x2f\xdb\x35\x30\xae"
+	"\xe8\x07\xa8\x4f\xf1\x9c\xbe\xd6"
+	"\xae\x10\xc6\xf2\x5e\xce\x03\x97"
+	"\x48\x1f\xa0\xbe\x01\x02\x81\x81"
+	"\x00\xea\x96\x35\xc0\x4e\xf5\xa8"
+	"\x4a\x4c\x6e\xe7\x88\x7b\xe3\x39"
+	"\xbf\xdb\x83\x23\xab\x36\x1c\xd1"
+	"\x0c\x00\xac\x55\x67\xb4\x5e\xa7"
+	"\xa9\x56\x41\xb4\x95\xbf\x5e\xa7"
+	"\x08\xfc\x8c\x40\x21\x5d\x53\x80"
+	"\x9a\xbb\x19\xbd\x9f\x11\xa8\xad"
+	"\xa0\xa0\xc6\xa7\x8d\xa7\x77\x1b"
+	"\xf8\x61\xda\xdd\x94\x4d\x64\x96"
+	"\xf2\xdd\x06\xbb\xc6\x4b\x09\x4d"
+	"\x66\x47\xbb\x8a\xcd\x5c\xd7\x4f"
+	"\x14\x9d\x06\xc4\xe5\xa5\x6b\xbb"
+	"\x19\xbc\x5a\x13\x6a\x2c\x63\x2d"
+	"\x62\x2f\x95\x1a\x22\x6b\x06\x00"
+	"\x30\x92\xf5\x08\xdb\x26\xdf\xd0"
+	"\x4d\x6a\xb8\x66\x94\xab\x56\x59"
+	"\xe9\x02\x81\x81\x00\xe2\x5c\xee"
+	"\x4d\xf3\x7e\xa8\x7d\xc8\x50\xbd"
+	"\x24\x7d\x55\x57\x2f\x92\x96\xdd"
+	"\x50\xfb\x93\xfd\x7c\x39\xdc\xcc"
+	"\x5c\xe6\xb9\x45\x15\x93\xf2\x0f"
+	"\x1c\x4a\x15\x39\x5e\x82\x8c\x7b"
+	"\x4c\x78\xd8\xdb\x6a\xc6\xde\x7c"
+	"\x04\x75\xf8\xa9\xd7\x9f\x59\x81"
+	"\x1e\x66\x16\x35\x62\xe7\xfc\xa3"
+	"\x73\x50\x7d\xd9\x34\xe0\xa3\x7a"
+	"\x5d\xbe\x60\xaa\x6a\xa3\x9a\x74"
+	"\x39\x3a\xbf\x37\x03\x9c\xa7\x81"
+	"\xe7\x89\x12\x22\x4f\x1a\xaf\xc0"
+	"\x10\xe8\xeb\x36\x06\x1a\xb3\xd0"
+	"\x1e\x8d\xb8\xcd\x6f\x96\x5e\x9e"
+	"\x7c\xb5\x25\x95\xf4\x49\xb7\x9e"
+	"\x22\xb0\xfe\xee\xc1\x02\x81\x80"
+	"\x52\x88\x10\x57\x1d\x30\x7e\x49"
+	"\xcd\x84\x43\x3f\x18\x5f\x04\x05"
+	"\x44\xe4\x69\x5b\x26\x09\x93\x5e"
+	"\x86\xaa\x1a\xce\x06\x7e\x3a\xbc"
+	"\x6f\xd5\x64\x04\x3f\x34\xbc\xcd"
+	"\x83\x94\x90\x47\x6c\xbf\x21\x02"
+	"\xcf\x0a\xaa\x54\xd9\x70\x9f\xcd"
+	"\x60\x63\x23\xb5\xc9\x92\x35\x2b"
+	"\xf8\xb9\x78\xb5\x61\x59\x1f\x66"
+	"\x48\x6e\xc3\x07\x6b\x84\xd5\x4f"
+	"\x43\xb3\x6e\x8c\xfb\xb9\xe0\x5f"
+	"\x9f\x33\x2f\x78\x9f\xef\x3b\xff"
+	"\x3c\xed\xef\xe0\xad\xc7\x22\xc3"
+	"\x65\x7c\x2e\x9a\xa7\x4e\x90\x28"
+	"\x86\xbb\x60\xa7\x41\x91\x02\x87"
+	"\x0a\x21\xe0\x2d\x33\xee\x88\x69"
+	"\x02\x81\x80\x01\xe0\x21\x3d\x56"
+	"\x1a\x23\x8d\x3f\x3e\xa4\x02\x8c"
+	"\x63\x4b\xa0\xa2\x47\x4b\xb8\x23"
+	"\x0d\x7b\xd9\x2e\xa5\x8c\xaf\xec"
+	"\x5d\x76\x9c\x12\x28\xbc\x45\x50"
+	"\xbb\xd2\x48\xdd\x2e\x0a\xcb\x7c"
+	"\x68\xd7\x90\x8f\xeb\x41\xfd\xcb"
+	"\x78\x6e\xf7\xd1\x63\x2d\x75\xb2"
+	"\x04\xd7\x16\x65\xb3\xab\x2b\x06"
+	"\x6c\xb4\x83\x02\xa5\x78\x1b\x14"
+	"\xbb\xba\x4d\xbd\xe3\x46\xb9\x3e"
+	"\xaa\x5c\xeb\x53\xbd\xf3\x8e\x96"
+	"\x8d\x26\xe9\x9e\xdf\xfb\xe8\x67"
+	"\xf5\xa2\xf1\xf4\x00\xaf\x65\xc1"
+	"\x68\xd9\x92\xe6\x1d\x0c\x95\x1d"
+	"\xef\x1c\xc9\xf4\x5e\x68\x1a\x03"
+	"\x7b\x69\x41\x02\x81\x80\x24\x8b"
+	"\xec\x97\xcd\xa4\x26\x85\x89\x66"
+	"\x6f\xf5\x6a\xee\xf0\x12\xd9\x15"
+	"\xda\xc5\xee\x1b\xfb\x18\xd1\x97"
+	"\x92\x46\x67\x5e\xc7\x76\x9c\xac"
+	"\xf3\x18\xb0\xaf\x38\x7e\x5a\x32"
+	"\x20\x34\xb4\x07\xa7\x66\x63\xcf"
+	"\xb5\x91\x6f\xbf\xe4\x4a\x62\x1f"
+	"\x19\x56\x8b\x53\xd3\x0e\x3b\x75"
+	"\xae\xbe\x5f\xac\x3f\xd3\x8c\x74"
+	"\xd2\x73\x46\x5f\xe0\x6a\x7a\xa6"
+	"\x3c\xe6\x10\xb4\x40\x1d\x5c\xda"
+	"\xea\x50\x3a\xa9\x59\xb8\x93\x0a"
+	"\x20\xfd\x4b\xd6\x67\x6d\xd0\xfe"
+	"\xc5\x39\xd6\xb9\x26\x84\x0c\xa4"
+	"\x78\xbb\x5f\x62\x68\x7d\xa9\xf5"
+	"\x01\xf9\x0c\xb4\x4b\x35\x00",
+	.key_len = 1447,
+	/*
+	 * m is SHA224 hash of following message:
+	 * "\x0a\x88\x8a\x69\x68\x9a\xc5\xaa"
+	 * "\xaf\x2d\x85\x55\x35\x28\x96\xed"
+	 * "\x38\xa6\x2d\xc6\x43\x89\x17\x0f"
+	 * "\x97\x05\xd1\xcc\x98\xd8\x36\xa1"
+	 * "\x50\x2c\x18\x47\x4a\xe6\x68\x32"
+	 * "\xbf\x4d\x09\xec\xc7\x58\xe3\xf4"
+	 * "\x80\xe6\x59\x14\x8b\xa7\x99\x22"
+	 * "\xcf\xd5\x4f\xd0\xcc\x85\xf8\xb9"
+	 * "\x19\x4d\xad\x19\x1c\x17\xd4\x91"
+	 * "\xbb\xb8\x6d\xdc\xac\xe6\xe4\xe7"
+	 * "\xd5\x9d\x51\x90\x91\x50\xee\xfe"
+	 * "\x45\x34\x5e\x1e\x77\x24\xea\x2d"
+	 * "\x2a\xb5\xd3\xac\x3f\xb8\x91\xd2"
+	 * "\x77\xea\x65\x7a\xa5\xdf\x56\x2e"
+	 * "\xf4\xc7\xf4\xb7\xb2\x22\x6e\xed"
+	 * "\xf9\xed\xd2\xad\xea\x4b\x7c\x21"
+	 */
+	.m =
+	"\xcc\x72\xea\x85\xea\x88\x07\x8c"
+	"\x25\xea\x3d\x04\xe1\x9c\x46\xef"
+	"\xf3\xfc\x8b\x34\xd5\x41\x0d\x06"
+	"\x24\xc5\xfc\xbc",
+	.m_size = 28,
+	.c =
+	"\x28\xb4\x79\x27\x5c\x8b\x50\x52"
+	"\xa3\x5b\x74\x39\xfd\xd0\xd9\x73"
+	"\x3d\xf3\xe5\xdd\x67\xcd\xca\x5b"
+	"\xba\x42\x7d\x29\x3c\x0b\xe8\x1d"
+	"\x7f\x08\x3d\xc1\x8a\x79\xe1\x7d"
+	"\xcd\x67\xa9\x6e\xe2\xd4\x2a\xc0"
+	"\xf7\x13\xe3\x47\x10\xf5\xbb\xee"
+	"\xfc\x60\x29\x3d\x53\xe0\x86\x39"
+	"\x04\xee\x83\x05\x64\xaa\xf6\x39"
+	"\xcf\xf7\xa4\xc4\x6c\xba\x22\xea"
+	"\xbb\x2d\x0b\x05\x52\x9e\xcf\x79"
+	"\x8c\xa0\x28\xf0\xc9\x48\xc7\x31"
+	"\x49\x2c\x1e\xa4\x38\x43\x31\xf7"
+	"\xa2\x85\xdb\xcc\xdc\xc4\x19\x41"
+	"\x2f\x62\x8f\xd3\xd3\xe1\x6f\xe9"
+	"\x66\xd6\x02\xbe\x34\x25\xca\xf5"
+	"\x74\x69\x85\xbc\xa7\xa2\x1a\x9f"
+	"\x98\xbc\xfd\x78\xe7\x18\x6b\xe3"
+	"\xaf\x52\x3a\x4d\xf3\x62\x8a\xc0"
+	"\xff\x79\x9e\x67\xf4\x85\x28\x8c"
+	"\xbd\x60\x71\x66\x1f\xb2\xaf\xac"
+	"\x6a\x17\xb9\x40\xd1\x10\x3c\x50"
+	"\x2e\x0c\x14\x3e\xf9\xae\x8f\x1f"
+	"\x4c\xbf\x6d\x60\x5d\x44\x6a\xd1"
+	"\x80\x37\x6d\x13\x80\xf3\xdb\x96"
+	"\x15\x2f\xfb\xba\xef\x20\x2c\xe1"
+	"\x8e\x1f\x9e\x48\xf5\xe2\xeb\x0c"
+	"\x28\x90\xa5\xcd\xcf\x43\x29\x7b"
+	"\x9e\xa5\xcd\xa9\xf9\x2a\x32\xe3"
+	"\x92\x3d\xc3\x28\x8e\x02\x42\xfd"
+	"\x9b\x26\x66\xa9\x77\x8c\x2c\xb1"
+	"\x6f\xde\x43\xaf\xdc\xb3\xae\x6c",
+	.c_size = 256,
+	.siggen_sigver_test = true,
+	}
+};
+
 static const struct akcipher_testvec pkcs1pad_rsa_tv_template[] = {
 	{
 	.key =
@@ -1490,6 +2210,845 @@ static const struct akcipher_testvec pkcs1pad_rsa_tv_template[] = {
 	}
 };
 
+static const struct akcipher_testvec pkcs1pad_rsa_sha384_tv_template[] = {
+	{
+	.key =
+	"\x30\x82\x08\x62\x02\x01\x00\x02"
+	"\x82\x01\x81\x00\xdb\x3b\x5c\x32"
+	"\xa2\x13\x31\xe2\xb4\x90\xf6\x82"
+	"\x58\x9a\xe4\xe3\x29\x0d\x2f\xac"
+	"\xdd\x07\xa6\x92\x77\xb9\x2f\x3a"
+	"\x50\x4d\x59\x27\xdb\x49\x7b\x91"
+	"\xb7\x6e\x84\x3d\x16\xe9\x9d\xe6"
+	"\xe2\x4c\x7c\x77\xd0\xed\x87\x63"
+	"\x5c\x12\x74\xfb\xac\xe8\x2e\x53"
+	"\x5b\x2d\xf7\xc1\x3a\x51\xbb\x23"
+	"\x83\x1a\x64\x13\x5b\xd9\x41\x43"
+	"\x5f\xb5\x5f\xac\xaf\x13\x18\x7a"
+	"\x51\x6c\xeb\x2f\x1e\x1e\x82\x11"
+	"\x88\x1c\x01\x5d\xc6\x19\xd8\x3a"
+	"\xf6\x78\xe8\xe2\xf3\x43\x05\x4e"
+	"\x9d\x8f\x80\x58\x90\xf1\x96\xf1"
+	"\x54\xc3\xd7\xd3\x48\x2e\x5e\x12"
+	"\xfe\x5b\x04\xb5\x86\xee\x34\x4b"
+	"\x3f\x7b\x13\x8e\x76\x14\x19\xff"
+	"\xea\xfe\x33\xed\xa7\xf7\x2c\x46"
+	"\x77\x7d\x8b\x75\x58\xc3\xca\xf4"
+	"\xe0\x94\x2f\x21\x72\x16\xe7\xd8"
+	"\xce\x2a\xec\xb2\x04\xd1\xa4\x1d"
+	"\xee\x49\x65\x60\xd5\x12\xfe\x59"
+	"\x57\x55\x00\x1d\x81\xdc\x23\xa9"
+	"\x10\x54\x99\x26\xfd\x55\xb7\xb9"
+	"\xdc\xe4\xfc\x76\x8d\x1e\x2e\xf1"
+	"\x1e\x87\xab\xe2\x4f\x25\xe5\x0d"
+	"\x3a\x9d\xc4\x94\x94\xa8\x4d\x04"
+	"\x1a\x28\xed\x3b\xa8\x6d\x5f\xfd"
+	"\x4c\xf1\x1a\x9c\x07\xf1\x20\x55"
+	"\x37\xd7\x60\x2b\xd5\x41\x12\xb9"
+	"\x0d\x4a\x49\xc6\xba\x44\x61\xc0"
+	"\x99\x2e\x10\x63\x7e\xb4\xaa\x74"
+	"\xfe\x8d\x03\xf4\x54\x09\xb3\x88"
+	"\xb0\x53\xc2\x8b\xad\x5a\x92\x3f"
+	"\x94\x7c\x4b\xfd\xfe\xb5\x1d\x65"
+	"\x2f\x80\xc3\x59\x56\x15\x40\xda"
+	"\x17\xbb\x11\xae\x5e\xc0\xed\xd4"
+	"\x3d\x20\xac\x7a\x04\x27\x24\xb7"
+	"\xe8\xc2\x8f\x69\xb2\x67\x5e\x43"
+	"\x65\xc4\x5a\x6f\xdc\x93\xf9\x64"
+	"\x2a\x4e\x2b\xc9\x7f\x36\xe4\x4c"
+	"\x35\x00\xc6\xaf\x43\x21\xd7\x0a"
+	"\x69\xe7\x0f\x36\xf1\x3d\xc4\xcb"
+	"\xa9\x26\xf3\x2b\x18\x21\x0d\xfd"
+	"\xc1\x03\xa6\x95\x9f\x8b\x42\x2a"
+	"\x30\x64\x7d\x54\x9e\x30\x86\xa8"
+	"\x97\x0e\x54\x8b\x96\x52\x6f\xde"
+	"\x1d\xb5\xfb\x87\x02\x82\x01\x80"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x01\x00\x01"
+	"\x02\x82\x01\x80\x6f\xb2\xc3\xdb"
+	"\xec\x66\x19\x41\xa7\xbc\x40\x4c"
+	"\xd3\x36\x40\x2e\x59\x29\x25\xc4"
+	"\x4e\x6d\x6a\x7a\x2f\xd3\xc3\x73"
+	"\x53\x9c\xc5\x33\x20\x70\x45\x2c"
+	"\x4e\x28\x59\xea\x02\x35\x8f\x45"
+	"\x26\xc1\xcf\x9d\x00\x6f\x35\x4e"
+	"\xf3\x87\x57\xe1\x2f\x0c\xb1\x25"
+	"\xb9\x4c\x27\x72\x64\xf3\x65\x84"
+	"\xc7\x74\x77\xce\xc1\x3a\x65\x1d"
+	"\x1d\xb3\x7c\x86\xb0\x40\xeb\x74"
+	"\xce\xed\xce\xba\x17\xff\x75\x47"
+	"\xd7\x70\x65\x65\xd0\xdc\x34\x26"
+	"\x7e\x93\xc6\x1c\x73\xd3\x0e\xee"
+	"\x2a\x2b\x06\x3a\x99\xc6\xed\xa0"
+	"\x12\x0f\x22\x74\xbc\xcf\x1b\x11"
+	"\x54\x55\x2d\x86\x0b\xb4\x6c\x22"
+	"\x76\xba\x8f\x91\x43\x2a\x24\xc1"
+	"\x16\x8a\xe4\x0c\x4f\x4b\x09\x5c"
+	"\xf1\xde\x9e\x62\x70\x92\x79\x62"
+	"\xe0\xd5\xbf\x59\xac\x16\x6d\x94"
+	"\x32\x1e\xfe\xb0\x08\xf2\x83\x4a"
+	"\xd8\xb0\xbe\x9d\x62\x92\xa5\xd3"
+	"\x01\x43\x06\xa7\x02\x16\x16\x79"
+	"\x3a\x1f\xd0\x53\x72\xcf\xa9\x57"
+	"\xe9\x0e\x2f\x89\x73\xfd\xa5\xc5"
+	"\x40\xc2\x55\x3d\x8c\x87\x49\xee"
+	"\x14\x82\x31\xec\x5c\x59\xd6\xc9"
+	"\xd0\xef\x16\x74\x56\xb9\x0d\x3b"
+	"\x86\xb2\x65\x5f\xf1\x50\x37\x4c"
+	"\x7b\x75\x91\x6d\x0e\xe8\x32\x91"
+	"\xb2\xd7\xa9\x8b\xfe\x2d\xca\x8e"
+	"\x04\x71\x3b\x50\x70\xd8\x1e\x02"
+	"\xf2\x9f\xe7\x16\x73\xaf\xeb\x3e"
+	"\x63\x24\xf0\x19\x9b\x87\x5c\x43"
+	"\x6b\x84\xd0\x4b\x74\x2e\xbd\xf6"
+	"\x64\x2a\xac\x6b\x72\x10\x8d\xfd"
+	"\x95\xb0\x2a\x1a\xb5\xfe\x71\xcb"
+	"\xf0\x1f\xcb\xb5\xf8\x3b\x53\xc1"
+	"\xcb\x48\x23\x77\x48\xc4\xa8\xaa"
+	"\x1e\x32\x6b\xee\xb4\x48\xfe\x47"
+	"\xf5\x71\x6b\x8e\xbb\x07\x77\x57"
+	"\x8b\xd6\xc4\x02\x12\x66\x23\x93"
+	"\x67\xbe\xcd\xd5\x0e\xd6\x29\xb6"
+	"\x57\x2e\xe3\xe8\x13\xa5\xbb\xcb"
+	"\xe3\xef\xf2\x0e\xdc\xf8\x5d\x08"
+	"\xb6\x9c\xb0\x73\xf6\x16\x40\x0e"
+	"\xbe\xea\x15\x41\xd4\xe2\xa1\xa5"
+	"\xce\xa9\xdf\x21\x02\x81\xc1\x00"
+	"\xee\xc0\x09\xbb\xd0\x11\x81\x81"
+	"\x37\xce\x0b\xa6\xd6\xc7\xf6\x58"
+	"\x45\x5e\x94\x63\xf0\x8e\x15\xeb"
+	"\x5d\x72\xd9\x10\xfd\xb9\xe1\x30"
+	"\xe0\x04\xde\x70\xb7\xbe\xb9\x68"
+	"\xc8\xbb\x1d\x9b\x58\x35\x08\x8f"
+	"\xa0\x10\x26\x30\xee\xce\xf8\x7b"
+	"\xf5\x20\x4c\x69\xab\xf0\x47\x6d"
+	"\x9a\xa9\x3f\xe3\x05\xeb\xdb\x15"
+	"\x4e\x4f\x9d\x8f\x67\x94\xda\x02"
+	"\x1a\x59\xe2\xc8\x23\x7e\xc1\x96"
+	"\x6f\x1e\xd3\x86\x04\xa1\xdd\xcb"
+	"\x0f\x0e\xa4\x14\x77\xae\x95\xba"
+	"\x1d\x17\x40\x1d\x5b\xab\x49\x4d"
+	"\x96\x6c\x97\xd8\x79\x49\x39\xf9"
+	"\xf9\x41\x38\x09\x5f\x2f\xad\x6e"
+	"\x84\xbc\xba\xfd\x35\x45\x3a\xa8"
+	"\x6c\x50\x2a\x3c\x99\xd8\x07\xc4"
+	"\xe0\xae\xf3\x13\x62\x80\xa7\x05"
+	"\x5c\x6f\xeb\x53\xaf\x9d\x90\xf9"
+	"\xf6\x0c\x81\x52\x24\xc2\x8b\xa5"
+	"\x8d\xc5\x69\xeb\x8f\xe0\x79\xc4"
+	"\x12\xd0\x1b\x31\x03\x2d\xe9\xfd"
+	"\x1e\x7e\x71\x41\xa0\x2e\x6c\x51"
+	"\x02\x81\xc1\x00\xeb\x12\x4f\x16"
+	"\xd1\x7a\x26\x8e\xd9\x12\xa3\x1a"
+	"\x63\xc8\x18\xb7\x8e\xb0\xbb\xa7"
+	"\xca\x7a\x1b\xcc\xe6\xf7\x9a\xb3"
+	"\x32\x13\xe4\xec\x54\x46\xf4\xed"
+	"\xd9\xea\x12\xfb\x46\x43\x5e\x7b"
+	"\xdf\xe1\x0c\xa1\xe2\xe3\x00\xd5"
+	"\x82\xf9\x7a\x77\x10\x1b\xa2\xfb"
+	"\xe6\x4b\x1a\x9f\x16\x38\x17\x6a"
+	"\x91\x25\x23\x16\x76\x57\x3b\x41"
+	"\x37\xda\x79\xdf\x1d\x67\x21\x43"
+	"\xff\xcd\xf8\x50\x14\xe8\xfd\x10"
+	"\xbb\x5f\x72\xb3\x10\xff\x92\xeb"
+	"\x4d\x4f\xaf\x4a\xa3\x09\x6e\x17"
+	"\xc6\x05\x76\x43\xf9\x3e\xe6\xb7"
+	"\xe5\xd5\x63\xe2\x8d\xc7\x5a\xbe"
+	"\xda\x5c\xd0\xa1\xb0\xf6\x15\x2b"
+	"\xbb\xd6\x92\x4a\x5e\xbb\x40\xe9"
+	"\x3d\x87\xd4\x6d\xff\x0e\x3b\xed"
+	"\xcd\xa9\x04\xd3\x7f\x63\xa3\x52"
+	"\x82\x1a\x3b\x03\xe1\x3a\x92\xb6"
+	"\xf1\x9f\xeb\x77\x80\x8a\x44\x36"
+	"\xf6\x86\x94\xef\xb1\x31\xed\x76"
+	"\xdd\x66\xe5\xab\x6c\x9b\xad\xa5"
+	"\xae\x2d\x6c\x57\x02\x81\xc0\x61"
+	"\xb2\x8f\xc9\x1d\x11\xac\xce\xf5"
+	"\xc9\x09\x17\x92\x98\x6d\x04\x1e"
+	"\x0e\xdf\xe6\x2d\x8f\xd7\x04\x63"
+	"\x4f\x15\xe8\x50\x0b\x7a\x4d\x07"
+	"\xbe\xb6\x4f\xa5\xc1\x14\x19\xfe"
+	"\xda\x22\xf9\x68\x94\xfb\x15\xda"
+	"\xc8\x00\xf8\xd1\x20\x3c\xaa\x6c"
+	"\xda\x74\xaa\x61\x43\x87\xd7\x20"
+	"\x0c\xd6\x29\xce\x48\x7f\x74\x2e"
+	"\x81\x81\xd6\x26\xea\xd0\x73\x3f"
+	"\xd9\x3c\x65\xed\xb0\x77\x03\x5a"
+	"\x21\x61\x29\x5c\x47\xe8\x7c\xa9"
+	"\x82\xb2\x3d\xf5\xe3\xa9\x3d\x69"
+	"\x9e\x14\x0c\x6f\x70\x9e\x24\x73"
+	"\xc3\xa8\x37\x5b\x1a\x4d\xf1\x2e"
+	"\xc0\x33\x70\x12\x55\x6e\x40\xe5"
+	"\x6c\x5d\x90\x33\xf6\x4a\x54\x95"
+	"\x4e\x10\x9a\x8f\xd1\xc9\x0c\x15"
+	"\x6d\x41\xe6\xd2\xce\x16\x8a\xa9"
+	"\x12\xdb\x57\x79\x6b\x9e\xe1\xca"
+	"\x7f\xbf\x83\x9f\xab\x44\x7d\xd0"
+	"\x8f\xd2\x12\x24\xe4\x14\xaf\xd9"
+	"\x8a\xf3\x35\x91\x39\x7e\xa0\xa7"
+	"\x93\xd2\x40\xaa\x8d\x94\xa1\x02"
+	"\x81\xc1\x00\x95\xae\xd3\xe3\xa0"
+	"\xc1\x9a\x39\xcd\xa9\x9b\x68\x1b"
+	"\x6f\x77\xa5\xb2\x5b\x81\x32\x87"
+	"\xb0\x97\x71\x33\x05\x29\x61\xbf"
+	"\x9a\x81\x63\xa7\x0a\x83\xbd\xcc"
+	"\xcb\xcb\x08\x6b\xa6\x4c\x4c\xae"
+	"\xb3\x3b\xfa\xfd\x61\x34\xcf\xc0"
+	"\x15\xa1\xd0\x9c\xd6\x2c\x43\x20"
+	"\x52\x64\x82\x36\x80\xc3\x10\x5d"
+	"\x99\xfb\x2c\xbb\x70\x32\x50\x06"
+	"\x03\xbd\x24\xc0\x31\x42\x32\x6d"
+	"\x79\xb7\x0b\x2f\x35\x68\xa7\x35"
+	"\xd8\xc2\x4d\x99\x9e\x53\xec\x37"
+	"\x0f\x6d\x7c\x26\x32\xc4\x02\xa9"
+	"\x5a\x23\x03\x87\x7b\x71\x28\x5a"
+	"\x56\x35\xc3\x36\x2a\x61\xf3\x33"
+	"\x52\x74\x9a\xf0\xf6\xd1\x66\x46"
+	"\xb6\x8b\xf2\x85\x3d\xd1\x48\x73"
+	"\xa0\xc1\x38\xc5\x77\x47\xdb\xd5"
+	"\xe0\x31\x2d\x87\x69\x67\xbf\x0b"
+	"\x0f\x2e\xd2\xfa\xa0\xf4\x79\x3f"
+	"\x84\x68\x48\x70\xf6\x1a\x17\xb1"
+	"\x84\xe2\xe3\x79\xbe\x64\x0a\x13"
+	"\x53\x09\x1d\xce\x0b\x97\x57\xef"
+	"\x80\x07\x03\x02\x81\xc0\x26\xb1"
+	"\x98\x36\x17\xa1\x6e\x69\x48\x2b"
+	"\x58\x51\xbe\x3a\x78\x12\x0b\x4d"
+	"\xf6\x36\xe3\x79\x67\xab\x50\xe9"
+	"\x43\x11\xff\xa1\x5b\x70\x05\x61"
+	"\x96\x89\x83\x3a\xec\xcb\xaf\x12"
+	"\x81\x4c\xfc\xe5\xc1\x4c\x51\xdc"
+	"\x5d\x06\xbe\x97\xbe\x38\xf7\xf6"
+	"\x47\xd7\x3b\xfe\xbf\x03\xb2\x2e"
+	"\xad\x82\xec\x37\xf9\x14\xb1\x28"
+	"\xf9\xff\x3b\x8f\xd6\x0a\xd6\xbf"
+	"\x5c\xa0\x13\x39\x08\x9f\x1b\x79"
+	"\xcf\xc9\xb5\x83\xe8\xb1\x20\x62"
+	"\x27\xbe\x36\xa3\xec\xde\xb7\x70"
+	"\x9b\x48\x72\x23\xb0\xb1\xd0\x06"
+	"\x49\x65\xf0\x46\xe4\xe1\xc7\xa1"
+	"\xa7\x43\x13\x26\xe8\xd8\x9d\x41"
+	"\x03\x6c\xf6\x6f\x3a\x91\xd9\xef"
+	"\x4d\xca\xa3\x2e\x48\x04\x60\xbf"
+	"\xb3\xb8\x5e\xa8\x96\x1b\x56\x60"
+	"\xd6\x89\x9a\x13\xd4\xad\x5c\x5a"
+	"\x5f\x0c\x17\x7c\x78\x2c\xa5\x13"
+	"\x8b\x6f\x50\x21\x13\x8b\x71\x9d"
+	"\xa3\xe9\xf2\x83\x9e\x23\x40\x04"
+	"\x09\x29\x1a\x1f\x4f\x72\x00",
+	.key_len = 2151,
+	/*
+	 * m is SHA384 hash of following message:
+	 * "\x56\xff\xeb\x36\xf2\xcc\xff\x84"
+	 * "\x8c\xf5\x8c\xa5\xc8\xc8\xaa\x67"
+	 * "\x77\x6c\xd5\x95\xcc\x37\x82\xf2"
+	 * "\x8d\xd8\x7c\x1a\x51\xf3\x0d\x71"
+	 * "\x1e\x0f\x89\x90\x36\x49\x31\x71"
+	 * "\x3c\xd0\x58\xfc\x58\x3f\x85\x49"
+	 * "\x6b\xb9\x8e\x07\xd8\xfe\xbd\xe0"
+	 * "\x8f\x22\x80\x24\x3e\x3c\x54\xe5"
+	 * "\xf1\xe4\xe4\x4d\x12\x16\x21\x06"
+	 * "\x73\xae\x93\x89\x17\x12\x14\x0e"
+	 * "\xfc\xba\x56\xa6\xd9\xfe\x56\x03"
+	 * "\x5d\x9c\xa6\x57\x50\xa5\x1f\xd0"
+	 * "\x5a\x70\x0e\x6d\x4d\x82\x04\xc2"
+	 * "\x60\x68\x06\x2e\x02\x1a\x7c\x36"
+	 * "\x79\xd0\xca\x5e\x0c\xaf\x55\x00"
+	 * "\x51\x31\xd0\x17\xd8\x82\x6f\x79"
+	 * "\xdd\x95\xa4\x93\x48\x8e\xde\xc7"
+	 * "\x19\x63\x55\x2f\xff\xa7\x35\x2d"
+	 * "\xb3\x7f\xd9\xbf\x1f\xfd\x66\xfe"
+	 * "\x03\x57\xfb\x2b\x49\x1d\x25\xb6"
+	 * "\x97\x05\xd6\x37\x48\x30\xb5\x53"
+	 * "\x90\xea\x5b\x72\xad\x59\x57\x4d"
+	 * "\x5d\x19\x34\xfe\x38\x06\x7f\xb4"
+	 * "\xcd\xed\xc3\xfe\xe2\x6e\xb1\x74"
+	 */
+	.m =
+	"\x3e\xea\x5d\xc1\x7d\x4a\x8e\x82"
+	"\x29\x12\x8f\xc4\x8b\xbd\xea\xfd"
+	"\x7d\x6e\x45\x5d\xcc\xf6\xf0\x36"
+	"\x70\xb7\x14\x02\x53\x50\xca\xcf"
+	"\xe6\xa6\x2e\xa3\xe4\x1e\x8f\xdd"
+	"\x30\x07\xde\x32\x97\x8c\x15\xdb",
+	.m_size = 48,
+	.c =
+	"\x71\xcc\x88\x5c\xde\x7e\x3c\x8c"
+	"\x06\xf7\x7b\x74\x4f\x9c\x9f\x7e"
+	"\xa8\x1f\x6f\xf1\xa7\xc9\x58\x6d"
+	"\x73\x13\x65\x30\xb6\x1f\x2b\xe2"
+	"\xe2\xfd\x97\x91\xf0\x9e\xb4\xb4"
+	"\x7e\x8c\x49\xa8\xd3\x06\x88\xd1"
+	"\x34\xeb\x3b\x04\x08\x86\x2b\xaf"
+	"\x40\x86\x9a\xeb\xfa\x54\xf7\x5a"
+	"\x7f\x10\x0c\x49\xe3\x51\x3f\x69"
+	"\xae\x13\x5d\xf1\xb2\x67\x6c\x32"
+	"\x01\x5a\x01\x0e\x22\xda\xa7\x36"
+	"\xc9\x46\xf5\x9f\xb8\x1b\x48\x83"
+	"\x89\xdd\x1f\xf2\x35\x21\x2d\x49"
+	"\x06\x23\xbe\x10\xaa\x94\x34\xf0"
+	"\x8c\xb2\x41\x81\x73\xba\x7d\x36"
+	"\x5c\x55\x17\x6a\x2c\x9b\x22\xab"
+	"\x2a\x3d\x1d\xa6\x09\x11\xe1\x71"
+	"\x72\x78\xfc\x97\x6e\x51\x9c\xea"
+	"\xd9\x0d\xcc\x52\x0d\x9e\xfc\xf6"
+	"\xad\x75\x21\xb7\x90\x1b\x93\xe0"
+	"\xff\x4a\x72\xb2\x19\x29\x39\x36"
+	"\xfd\x5d\xe5\x15\x91\xc6\x52\x34"
+	"\xd6\xfb\x36\xaf\xc8\x48\xf8\x62"
+	"\xa6\xf5\xbf\x0e\x39\x30\xb1\x2a"
+	"\xe6\xc1\x48\x67\xbc\x6e\x52\x34"
+	"\xed\x74\xbb\x2e\x67\x33\xac\xf7"
+	"\x28\x45\xd4\xee\xec\xb0\x69\xf1"
+	"\xb1\xdf\xf7\x94\xa4\x5b\x03\xc1"
+	"\x3c\x1a\x02\xc9\xbd\xf7\xcd\x19"
+	"\x1f\x05\x7b\xdb\x7f\x12\xc0\x98"
+	"\xf2\xc5\xb0\x14\x35\x8a\x53\x34"
+	"\xb7\x67\x64\xca\x58\x2c\xd3\x0f"
+	"\x28\x36\x3d\xad\xfc\x4f\xdd\x07"
+	"\x96\xa0\x2b\xf0\x3e\xd3\x4a\xd7"
+	"\x97\xfa\x7c\x42\x2e\xd3\xc9\x26"
+	"\x25\x3f\x70\xda\xd3\x8a\xa7\xd7"
+	"\x9e\x99\x8d\x71\x17\x0c\x86\x6d"
+	"\x2b\x5b\xb8\x22\xac\x31\x5d\x0e"
+	"\x0c\xf9\xe7\x74\xdf\x75\x1a\xfa"
+	"\x73\x3b\xa8\xdc\xd0\x86\x0f\x9f"
+	"\xe6\x42\xa9\xe6\xcd\x8e\xfa\x35"
+	"\x2a\x4f\xc5\xea\x70\x15\x22\xe1"
+	"\x5f\x70\x65\xb2\x60\x0e\x82\x7a"
+	"\x58\xef\x01\x7d\x16\x23\x33\xfc"
+	"\x29\x89\xeb\x43\xf1\x9c\xa7\xf4"
+	"\x28\x7d\x30\x8a\x93\x39\xfd\x53"
+	"\x97\x57\x81\x57\x70\xc6\x1f\x94"
+	"\x1e\xa5\xed\x69\xfb\x6a\xd1\xe6",
+	.c_size = 384,
+	.siggen_sigver_test = true,
+	}
+};
+
+static const struct akcipher_testvec pkcs1pad_rsa_sha512_tv_template[] = {
+	{
+	.key =
+	"\x30\x82\x0b\x28\x02\x01\x00\x02"
+	"\x82\x02\x01\x00\xb0\xf8\x80\x65"
+	"\x86\xb7\x22\xf4\x14\x07\x04\xf0"
+	"\x00\xf5\x1f\xc9\x03\x28\xad\x8e"
+	"\xc8\xe4\x0d\xeb\x69\x14\xf9\xcc"
+	"\x4a\xfe\x29\x63\x14\xb9\xcc\x6b"
+	"\xbc\x16\x5b\xf4\x27\x94\x4c\xa0"
+	"\xd6\x68\x6b\xfd\xab\x52\xfe\xd2"
+	"\x53\x3a\x3a\x86\xcd\x6a\x58\x4c"
+	"\xd8\x7a\xa3\xaa\x0f\xae\x92\xab"
+	"\x30\x8a\xc6\x96\x7b\xba\x04\xd4"
+	"\xc2\xea\xb4\xca\xfe\x7e\x84\x9a"
+	"\x69\x2e\xef\x11\x34\xd1\x0d\xa2"
+	"\x1b\x5b\x8f\xb2\x94\x54\xbb\xfa"
+	"\xd6\xff\xb0\x98\xec\xf1\x6c\xd1"
+	"\x61\x52\x1f\xcc\x92\x66\x7f\x95"
+	"\x0c\xd4\xe3\x4b\x54\x24\xb0\xc8"
+	"\xa2\x05\x00\xa9\x93\x88\x64\x82"
+	"\xa9\x35\xa8\xc6\x06\xbd\xc9\x10"
+	"\x88\x91\xf9\xe8\x60\x9a\xbe\x17"
+	"\xc3\x9e\xf4\x6b\x92\xf5\xb4\xe7"
+	"\xa3\x24\xcd\xa9\x42\xb8\xfd\x0e"
+	"\x73\x4b\x2a\x2b\xfc\x2d\x75\x48"
+	"\xf8\x8f\x1b\xad\xd8\xab\xa9\x5a"
+	"\x8c\xde\x63\x12\xe8\x89\xb4\x38"
+	"\x5e\xaa\xfe\xf1\x39\xae\xd6\xdf"
+	"\x76\x4f\xb6\xe3\x4a\x5b\xd3\xe1"
+	"\x4d\xe8\x2e\xe5\x9f\xd2\x56\x51"
+	"\x1b\x5f\x66\x15\x4e\x0d\xfa\xed"
+	"\xb5\x74\xdd\x40\x51\xdc\x72\x56"
+	"\x62\xd9\x21\xd2\x73\xc0\x34\x6d"
+	"\x67\x34\xb4\xcb\xfc\xc3\x83\x44"
+	"\xec\xba\x5c\x44\x66\xb5\xb7\xee"
+	"\x50\x3f\xa6\x79\x5c\xb5\x7d\x93"
+	"\xc1\x6c\x05\x34\xef\xb9\xcd\xf6"
+	"\xb1\x8d\x26\xec\x39\x12\x31\x5d"
+	"\x12\x68\xd2\x9c\xe2\x74\xc4\xe5"
+	"\xbf\x46\x6e\xdd\xb6\x92\xee\xd3"
+	"\x9f\x7b\x4b\xe4\x8b\xde\xf2\x3f"
+	"\xd4\x0e\x5c\x11\x40\x9a\x50\xac"
+	"\x81\xab\x9f\x60\xf5\xad\xa7\xa2"
+	"\xb9\xf9\x57\xe8\xc0\xf3\x24\x1f"
+	"\xa1\x89\x95\x31\x19\x32\x66\xd5"
+	"\x8a\xf5\xe4\x66\x65\xf8\xa1\x5a"
+	"\xfb\xb0\xc6\xef\x47\x87\x4a\x60"
+	"\xdd\x0c\xa5\xfe\xd5\x63\xc4\x39"
+	"\x2c\x2e\xd8\xb1\x7b\x53\xa6\x28"
+	"\xa1\x9f\x1b\xf1\xda\x28\xd9\x7e"
+	"\xdf\x15\xc2\x64\x61\x81\x55\x0a"
+	"\x79\x3d\xe6\x9f\x1e\x59\xc7\x74"
+	"\x16\xb0\x75\x48\x02\xb5\x76\x08"
+	"\x64\x10\x53\x6c\xba\xfa\x8e\x93"
+	"\x1b\x55\xb2\x40\x93\x4d\x30\x1b"
+	"\x1f\x11\xc0\x8a\x7e\x73\xe1\x45"
+	"\x46\x95\x5f\x2a\xde\x09\xe5\x11"
+	"\x1f\xa0\xb3\xa2\x29\x08\xac\x96"
+	"\x28\x3b\x8a\xa1\x1e\xb2\x22\x8c"
+	"\x5e\xc1\xe5\x26\x5b\x9f\x62\xa1"
+	"\xb2\x01\xa2\xba\xa0\x75\xfa\xd3"
+	"\xa4\x23\xd4\x26\x82\xeb\x39\x0a"
+	"\x0c\xc8\x0d\x8a\xc3\xda\x92\x62"
+	"\xd1\x5a\xe1\x3f\x4d\xd7\xf2\x70"
+	"\x32\x9f\x4e\xc9\x24\x18\x34\xfc"
+	"\x5e\xc6\x05\x00\xa8\x3d\x86\x02"
+	"\xf8\x10\xa4\x4b\x6b\xb9\x2c\x84"
+	"\x49\xec\x22\x45\x02\x82\x02\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x00\x00\x00"
+	"\x00\x00\x00\x00\x00\x01\x00\x01"
+	"\x02\x82\x02\x00\x2a\x65\x1d\xbd"
+	"\x34\xac\xa2\xbe\xfe\x1d\x40\x5d"
+	"\x8e\x9d\x2a\x64\x29\x51\x91\x62"
+	"\x67\x55\xda\xcf\xce\x78\xab\xee"
+	"\xf8\xf8\x32\x49\xfe\x29\x14\x83"
+	"\xaa\xa7\x60\x97\xf8\xf3\x59\xf8"
+	"\x1d\xe5\x1e\x19\xc2\xf5\x3a\xfe"
+	"\x7f\xd9\xaf\x33\xa3\xfa\x19\xf5"
+	"\x2e\xd7\x12\x66\x27\x23\x97\x17"
+	"\x53\xa5\xab\x35\x59\xa6\x92\x89"
+	"\xb5\xb8\x2c\x5f\x84\xe0\x22\x3c"
+	"\x2b\x7e\x86\xdf\x04\x5c\x20\x7b"
+	"\xf9\x39\xd2\x0e\xe1\xe1\xfa\xb6"
+	"\x78\xd3\x00\x27\x62\x7b\xe3\xa9"
+	"\xb5\x3a\xcb\x4b\x62\xb7\x13\x51"
+	"\x4d\xa5\xa2\x5d\xe5\x66\x74\xc3"
+	"\x5b\xac\xff\x1f\xa0\xac\x48\x07"
+	"\xd5\x29\x79\xa9\x16\xec\x58\x42"
+	"\x91\xb5\x06\xbd\x58\x87\x39\x8a"
+	"\xcf\xde\x40\x87\x94\xc9\x2f\xf1"
+	"\x5d\xef\xef\x24\xe1\x4b\x17\xb4"
+	"\x62\x82\x5b\x33\xf6\x51\x9d\x8b"
+	"\x41\xa0\x66\xed\x1a\xbe\xe4\xef"
+	"\x40\x57\xf6\x3c\x0d\x61\x0f\xbf"
+	"\x81\x4e\x4d\x8c\xd4\x58\xc2\x38"
+	"\xfe\x4a\xd2\x71\x58\xda\xee\xc3"
+	"\x46\xba\x9c\xa1\x7d\xcc\x11\xc9"
+	"\x94\x59\x22\x47\x70\x4f\x81\x11"
+	"\x79\x2a\xe9\xe6\x02\xb6\x60\xdd"
+	"\xb9\x08\x73\x45\xcb\x5c\xfc\x32"
+	"\x49\x4b\x23\x14\xcb\x14\xbb\x59"
+	"\xc3\xac\x51\x76\x41\x74\xbe\xff"
+	"\x64\x52\x1b\x3b\x09\x76\xa1\xaa"
+	"\xbb\x1f\xda\xe4\xb9\x8d\xd7\x7d"
+	"\x6b\xad\x3f\x87\xac\xe5\x5f\x6d"
+	"\x86\x66\xf1\x1a\x48\xf0\xba\xe8"
+	"\x9c\x49\xb7\x31\x1b\x07\xad\x28"
+	"\x41\x0a\x77\x75\x91\x13\x26\x82"
+	"\x7f\x4a\x9f\x62\x96\x9f\x80\x54"
+	"\xa4\xa5\x33\xd0\x6d\x84\x3c\x3f"
+	"\x9e\xd8\x65\x98\x63\xe2\x77\x76"
+	"\x8f\x89\x60\x68\xcf\x04\x63\x35"
+	"\xbd\xb8\xc0\x5b\xe8\x2e\xb6\xb0"
+	"\x5a\x06\x6c\x53\x02\x98\x1c\x51"
+	"\x7f\x6b\x85\x46\x66\x7a\x08\xc9"
+	"\x67\xca\x30\x40\xca\x5c\x4b\x85"
+	"\x68\x4b\x7b\x9d\x29\x50\x5b\x4e"
+	"\xac\xcb\x9b\xa4\xb9\xe2\x47\x29"
+	"\xdb\x08\x92\x20\xba\x9d\xf8\x11"
+	"\x95\x6c\xc6\x22\x57\xb8\xd6\x76"
+	"\x5e\x64\x78\x2e\x62\x46\x8c\xd8"
+	"\x41\xbc\x26\x2a\x84\xa4\x7b\x62"
+	"\x23\x2e\x78\x12\xb5\x7b\x9b\x89"
+	"\x1b\xae\x7a\x45\x4b\x3c\xf4\x15"
+	"\xac\xa4\xe8\xe3\xe0\x64\x80\x1a"
+	"\x31\x95\xaf\xc5\xd9\x55\x4d\x4e"
+	"\x68\x47\x0d\xa7\x5e\xdd\x5d\x70"
+	"\xe9\xc2\x0a\x32\x7a\x11\x62\x92"
+	"\x5e\x37\x06\x67\x4d\x68\x87\x65"
+	"\xaa\xdb\x41\x35\x6e\x36\xba\xb6"
+	"\x64\x7c\x79\x84\xbe\x35\x16\x43"
+	"\xdf\xf9\xd9\x16\xc1\x49\x62\x49"
+	"\x5f\x14\x0a\x90\x5c\x60\xaf\xee"
+	"\x09\xea\xce\x87\x71\x11\x24\xce"
+	"\xf3\x29\x61\x81\x02\x82\x01\x01"
+	"\x00\xe3\xe0\xd4\x64\xb3\xa8\x8f"
+	"\x16\xd0\x33\xfb\xfc\xca\xc5\x4c"
+	"\xbc\xb7\xc6\xe8\x6c\x71\x55\xbf"
+	"\xe1\xfa\x6e\x12\x48\x3d\x94\xae"
+	"\xb9\x98\xf4\xe1\x59\x24\xfd\x79"
+	"\x73\x6e\xc8\xad\x90\x83\x38\x02"
+	"\x8d\x27\x4c\x87\x56\x5c\xed\x66"
+	"\xbf\xa0\x51\x53\xfa\x7c\x87\x0b"
+	"\xe5\x79\xf5\xda\x3f\x2b\x11\xf5"
+	"\xb4\x1c\xed\x6a\x11\xc6\xfe\x73"
+	"\x2f\x86\x76\x20\x46\x0f\x9b\x84"
+	"\x76\x33\xaa\xc5\xb4\xfa\xe9\x16"
+	"\xbb\x10\xe5\x0c\x3f\x02\x68\x69"
+	"\x10\x81\x30\x4a\x90\x29\xd7\xfb"
+	"\x48\xc7\x98\x59\xb5\xbf\xae\xf6"
+	"\x76\x8f\x6b\x12\xaf\x0d\xc6\x2e"
+	"\x2c\xfe\x35\xdf\x62\x01\xb7\xee"
+	"\x49\xd5\x29\x49\x43\xe4\xfe\x3c"
+	"\x03\xeb\x7a\x04\x6a\xbc\xaf\x44"
+	"\xab\x7d\xe6\x65\xa3\x98\x9c\x77"
+	"\x03\x18\x4e\x90\x13\x3a\x47\x81"
+	"\x8b\x1f\x5d\x11\x6e\xfb\xe2\x56"
+	"\x52\x96\x38\x59\x48\xea\x4e\x9e"
+	"\x25\xe8\xa2\x20\x34\xa3\x90\x1f"
+	"\x01\xee\xa9\x73\xd9\xeb\xca\xbb"
+	"\x36\x76\xc7\x64\xeb\x74\xf7\x10"
+	"\x58\xd1\x77\x41\x91\x1e\xf5\x2b"
+	"\x86\x34\x3f\xe8\x34\x1e\x23\x01"
+	"\xf5\xd6\x36\x40\x60\x50\xec\xd5"
+	"\x22\x57\x75\x77\x75\x3d\xa5\x91"
+	"\xc5\x25\xf0\xc4\xe1\xab\x5b\xa0"
+	"\xb7\x48\xbb\x44\xff\x82\x2c\x40"
+	"\x5d\x02\x82\x01\x01\x00\xc6\xcf"
+	"\x64\x59\xf5\x21\xdf\x0d\x49\xcf"
+	"\x81\x4e\xb4\x73\x45\x65\xa2\x15"
+	"\x89\x33\xd8\x1d\x73\xa1\xa1\xbb"
+	"\x5e\x21\x34\x6e\xb5\xb9\x81\x54"
+	"\x9b\xcb\xc6\x8f\xf2\x23\xc0\x21"
+	"\xf3\x31\xcc\x27\xd5\xae\x43\xec"
+	"\xf2\x8e\x16\xc1\x58\x2b\xe9\x14"
+	"\x6c\x1e\x3e\x3a\x96\xe0\x3a\x7c"
+	"\x39\x19\x07\x00\x97\x8f\x84\xa1"
+	"\x91\x4e\xce\x3b\x2c\xc8\xe5\xfd"
+	"\x45\xad\xcb\x08\xf5\x9d\xf2\x38"
+	"\x80\x14\x5f\xef\xf4\xfb\x2d\xb1"
+	"\x47\x47\xf6\xf9\x1d\x12\xbf\x9d"
+	"\x8e\x0c\xe4\xd0\xe9\x9a\x0e\x5d"
+	"\x8c\xc7\xe1\xba\x89\x22\xa0\x56"
+	"\x8d\x4b\xcb\xb2\x77\xa8\x82\x63"
+	"\x21\x04\x7d\xbe\xe0\x9f\xa3\xbe"
+	"\xaf\xe0\xeb\x6c\xc6\x4a\x71\x1b"
+	"\xfa\xf2\x95\x7e\xf8\x44\x9e\x30"
+	"\xb8\x7c\xe5\x9b\xc7\xe9\x67\x58"
+	"\xe4\x53\xa6\x90\x2a\x65\x26\x68"
+	"\x7b\xee\x4e\xe1\x09\xbf\x58\x3c"
+	"\xc8\x76\xa1\x85\xd9\x71\x9a\xf9"
+	"\xeb\xb7\x0f\xb3\x14\x5e\xca\xac"
+	"\x8d\x4c\xd7\x94\xb2\x17\x1b\x13"
+	"\x04\xb9\x89\x7c\x64\x58\x2a\x1e"
+	"\x40\x28\xfe\x81\x7b\xcd\x64\x4b"
+	"\x24\x96\x6a\xe7\x8c\xd3\x49\xaf"
+	"\x51\xc6\xfd\xb7\x06\x00\x60\x7e"
+	"\x6b\x0a\x88\x06\x87\xa4\x03\x6c"
+	"\x94\x04\x3e\xcd\x02\x3c\xcf\x4d"
+	"\x1e\xa1\x69\xe3\x6b\x09\x02\x82"
+	"\x01\x01\x00\x84\x28\x48\x05\x7f"
+	"\x0c\x2a\xb0\x18\x5f\x8d\xac\xb6"
+	"\x0c\x66\xcc\xab\xe8\x77\xb3\x2c"
+	"\xa7\x86\x79\x10\x3b\xf5\xee\x5f"
+	"\x88\x73\x2c\xc0\x40\x84\xd0\xa6"
+	"\x5b\x3a\xac\x6f\x7b\x82\xed\x64"
+	"\xac\x7f\x0e\x4d\xd4\x5c\x49\x01"
+	"\xeb\xcc\xc3\x25\xbf\x04\xed\x28"
+	"\x7b\x49\xb2\xf4\x60\x59\x0a\x79"
+	"\x78\x18\xff\x34\x27\x9c\x74\x62"
+	"\xeb\xfe\xf1\x79\x11\xd8\xf5\xc8"
+	"\xe7\x70\x92\xfc\xb3\xf5\xa5\x58"
+	"\x3f\x06\x27\xc7\x4c\x8f\x25\x9b"
+	"\xc7\xe1\xe6\xe8\x85\x91\x38\xba"
+	"\x8b\xbf\xf8\x1d\x89\x85\xbd\x9a"
+	"\x45\x66\x3a\xac\x42\x74\xbb\x81"
+	"\x49\x86\x4e\x21\xc0\x2c\xa5\x32"
+	"\x49\xf1\xb0\x70\x27\x59\x8e\x3e"
+	"\xa8\x13\x36\x52\xc2\x98\xfd\x2e"
+	"\x20\x97\xba\xbd\x71\x79\x50\xbf"
+	"\x99\x3e\x8c\x54\xf4\x7b\x2a\x2b"
+	"\x8a\x7c\xde\xaa\x79\xd7\x3a\x7e"
+	"\x6a\x6d\x95\xc7\xf9\x36\x0e\xcb"
+	"\x94\x7c\xb3\x8f\x8f\x37\x0e\x50"
+	"\x2d\xad\x4d\xf9\x93\xdc\xd4\x6d"
+	"\xa3\x4d\x9f\xf0\x93\x03\x3c\x6a"
+	"\xc1\x3e\x81\xc4\x39\x5d\x95\x44"
+	"\x56\xfb\xbf\x01\x5e\x6e\xd4\x8b"
+	"\x05\xd2\x85\x38\x8e\x8e\x17\x4d"
+	"\xd7\x8c\x64\x36\x44\xb1\xdb\x14"
+	"\xbd\x89\x24\xc5\x34\x35\x7a\x8b"
+	"\x5e\xc0\x7e\x8b\x77\xd0\x34\x3e"
+	"\xd2\x5b\x21\x02\x82\x01\x01\x00"
+	"\xc4\x7c\x86\x1e\x8e\x31\xca\xa3"
+	"\xf4\x75\xfb\xc3\x50\x7e\xb8\x70"
+	"\x88\xef\xa5\xe9\x32\x47\x71\x57"
+	"\xa6\x63\x6a\xd5\x80\x5f\x86\x84"
+	"\xef\x78\x32\x7e\x04\x21\x91\xdb"
+	"\x99\x71\x34\x4c\xdc\x0b\xe6\x98"
+	"\xd3\x93\x23\xea\x20\x3d\x38\x2a"
+	"\x35\xb2\xb4\x64\x73\x76\x2e\x55"
+	"\x30\x18\x43\xca\x23\x6c\x87\x5e"
+	"\x15\x64\x25\x1b\x70\xa1\x75\x52"
+	"\x9c\x53\x63\x81\x2f\x30\x0c\xa7"
+	"\xec\x93\xc0\xa9\xe2\x63\x62\x88"
+	"\xb8\x91\x03\xf4\xf0\xf7\x25\xaf"
+	"\xb8\xe7\x3b\x54\x2c\xf4\xf8\x5b"
+	"\x04\xa5\x9a\x94\x95\xd8\x68\xde"
+	"\x6a\xfe\xff\xa8\x90\x8a\x32\xce"
+	"\xbd\x06\xc1\xc4\x11\x8f\x0b\xab"
+	"\x5b\x07\x04\x06\x54\x46\xed\xf6"
+	"\xf6\xb1\x03\xa4\xaa\xf1\x90\xe7"
+	"\x28\x3c\xd8\xa8\xbf\xb8\xe5\xac"
+	"\x1d\x32\xf2\x7e\x66\x19\x07\x7d"
+	"\x6e\xa7\x47\x0c\xb0\xb2\xa2\x92"
+	"\x94\xdf\x85\xf4\x11\x11\x1b\x65"
+	"\xb7\x18\x17\xc9\x95\xdf\x94\x5b"
+	"\x57\x0d\x90\x9d\x08\x57\x72\x67"
+	"\x79\x9f\xa3\x44\xc7\xc7\xc3\xd8"
+	"\x8a\xcf\xd8\xf3\x6b\x0d\x87\xd0"
+	"\xe7\xb5\x88\x54\x8d\x15\x89\xcd"
+	"\xec\x7e\x34\x61\x69\xb1\x70\x62"
+	"\x00\xfc\x10\x48\x99\x45\x72\xdb"
+	"\x3b\x73\x1d\x1a\xd1\x02\x86\x9a"
+	"\x98\xfb\x0c\x31\xd5\x58\x29\x19"
+	"\x02\x82\x01\x00\x00\x87\x87\x9b"
+	"\x3d\x45\x6c\xfb\x63\x20\xa1\xb9"
+	"\xf5\x69\xce\x6b\x96\xae\x69\xf0"
+	"\xe1\xf2\x8b\x39\x2a\x18\xc8\xca"
+	"\x58\xc1\x30\x2f\x15\xca\x93\x89"
+	"\x38\x37\x83\x85\x6e\x09\x5b\xad"
+	"\xac\x43\x23\xa3\x19\xfd\x1d\xc8"
+	"\xa2\x8f\x75\x55\xc8\xf7\x0e\x37"
+	"\xa8\x52\xd1\xb6\x94\x03\xc9\xe1"
+	"\xbf\xde\x4f\xd9\x3c\xa0\xc4\x2d"
+	"\x7d\x6a\x57\x02\xb6\x17\x05\xd7"
+	"\x0d\x43\xce\xf0\x2b\x95\xb3\x31"
+	"\xdf\x96\x03\x36\x5e\x5d\xed\xbd"
+	"\x71\xc3\xf2\x64\x2c\x3d\x3b\x64"
+	"\xfa\x3b\x83\x94\x6d\x07\xda\x52"
+	"\x44\x60\x51\xe5\x4b\xcf\x1d\xec"
+	"\x50\x50\xa6\xea\x39\x40\x18\x17"
+	"\x1d\xd8\xd2\xdf\xc9\x4e\x52\xf4"
+	"\x92\x95\x49\xd6\x3d\x1e\x03\xa2"
+	"\x85\xd7\x5c\xa4\x84\x6c\x80\x22"
+	"\x63\x14\x6c\x41\x20\x23\xbb\x54"
+	"\xc7\x05\x5f\xb7\xb4\x4e\xbd\x79"
+	"\x7d\xeb\xb6\x47\xcc\x70\xeb\x0e"
+	"\x98\x12\x09\x44\x7e\x22\x34\x6a"
+	"\x11\x17\x2f\xdd\xe2\xac\x51\xe9"
+	"\x97\x1c\xb1\xe9\xa2\x3e\xb2\xce"
+	"\x25\x9d\xb6\x67\xed\x11\x93\xa8"
+	"\x62\x49\xcb\x35\x41\xd2\xdd\x9b"
+	"\xb5\xb4\xcb\xcf\xd3\xd5\x09\xc4"
+	"\xfe\xfe\x49\x01\xc7\x0a\xe8\x9d"
+	"\xe3\x09\xa9\x3d\x73\x7f\x20\x8f"
+	"\xbb\x56\xda\x27\x28\x39\xbc\xf6"
+	"\x59\xfe\x77\xb2\x00",
+	.key_len = 2861,
+	/*
+	 * m is SHA512 hash of following message:
+	 * "\x15\x1a\x49\x51\x70\x2b\x73\xd2"
+	 * "\xe6\xd0\xdc\x0d\x75\x72\x3a\xfd"
+	 * "\x3f\x1c\xa8\x78\x3a\x3e\x1e\xfd"
+	 * "\x69\x6e\xf5\x77\x45\x92\xce\x9b"
+	 * "\x4b\xf7\x61\x12\xba\xbc\x18\xc4"
+	 * "\x0d\x53\x5f\x79\x4e\x18\xfb\x42"
+	 * "\x88\x7c\x78\x71\xae\x8a\xae\x76"
+	 * "\x2c\x03\x0d\x89\xef\x80\x25\x46"
+	 * "\x4a\x2a\x26\x70\xb5\xc3\x7a\xb7"
+	 * "\x8e\x57\x03\xe9\xb5\x4a\x01\x76"
+	 * "\xf7\xe5\xf8\x6b\xd8\x18\xc6\x71"
+	 * "\x68\x17\xd5\x34\x73\xa6\x1b\x10"
+	 * "\x3c\x14\xc4\x50\x8f\xf3\x28\x81"
+	 * "\xe8\x51\x8a\x75\xfc\x16\x8d\x01"
+	 * "\x37\x81\x30\x0b\x91\x62\x5b\x0f"
+	 * "\x7a\xb6\x64\xde\x44\x62\x08\x18"
+	 * "\xcc\xff\x00\x42\x4c\x39\xf8\x28"
+	 * "\xf0\x89\xce\x33\x28\x41\x28\xcc"
+	 * "\xf5\x19\x96\x7c\xbf\xe7\x10\x3d"
+	 * "\x57\x10\x89\x88\x1b\x19\x0b\xb6"
+	 * "\x84\xb2\xb6\x24\xe3\x82\x1b\xce"
+	 * "\xfe\xf0\xf3\xa1\x9f\xb2\x1f\x04"
+	 * "\x3f\x7f\xf7\x4f\x19\x8d\xcb\x83"
+	 * "\x24\xe3\xc5\x67\x08\xf6\x18\xe3"
+	 * "\x2b\x6d\xe6\x47\x46\x67\x4c\xec"
+	 * "\x69\xaf\x0a\x8b\x6a\x35\x20\x1d"
+	 * "\x77\x91\x7f\x34\xfd\x36\x24\x6b"
+	 * "\xfa\xfd\x8c\x2f\x6f\x85\x7e\x08"
+	 * "\x7b\x5e\xd6\xc0\xd0\x97\x95\xbe"
+	 * "\xb4\xf4\x86\x3d\xf8\xc6\x80\xb1"
+	 * "\x12\x35\xec\x86\xff\x94\x41\x02"
+	 * "\x2e\xb7\xde\xdc\x25\x00\xe0\xf0"
+	 */
+	.m =
+	"\x75\x6e\x38\x30\xab\x85\x52\xff"
+	"\xb6\xa2\xa7\x16\x22\x05\xb6\x63"
+	"\xaa\xc7\xdb\xdd\x37\x38\x25\x02"
+	"\x96\x64\xe3\x16\xf7\x81\xfc\x9b"
+	"\x31\x1b\x52\x8f\x51\x21\xc1\x18"
+	"\xb7\xd3\xc6\xd6\x1a\x9e\xd0\x83"
+	"\xcd\x5b\x02\x60\x61\x67\x55\xb5"
+	"\x56\x37\x34\x6a\xbd\x22\x46\xb1",
+	.m_size = 64,
+	.c =
+	"\xae\x51\xd3\xbf\x63\xc8\xaa\x4d"
+	"\xce\xea\xa9\xf3\xea\x3b\x89\x5d"
+	"\x76\x00\xcf\xc5\xe8\xdc\x06\xe8"
+	"\x2e\xb7\x84\xca\xaf\xdf\xe9\x36"
+	"\x7b\x02\x04\x17\xf0\xe3\x88\x42"
+	"\xb2\x89\xe2\x2f\x2e\x26\xd2\xeb"
+	"\x4e\x6a\xb8\x70\x94\x33\x96\x0b"
+	"\xa1\x73\x9d\x47\xdd\xa2\x42\x5b"
+	"\xb4\x4b\x9c\x54\x65\xbc\x45\x4b"
+	"\xa8\x43\xde\x8c\x30\xf0\x00\x1c"
+	"\x38\x40\x07\x9a\x7c\x99\x6d\xfe"
+	"\x4e\xf0\x56\x53\x56\x8d\x60\x0e"
+	"\x91\x5a\x8e\xc2\xc1\x78\x94\x85"
+	"\x31\x6a\x74\x90\x65\xcd\xa2\x74"
+	"\xaa\x41\xcd\xf7\x0f\x98\x6b\x29"
+	"\x09\x0b\x90\x71\x2a\x97\x14\xd5"
+	"\xa9\x03\x2d\x10\xb8\x8a\x7a\x54"
+	"\x16\x08\xb2\xe7\x6a\xd3\x9a\xb5"
+	"\x24\x71\x59\xe5\x35\xce\x47\x58"
+	"\x8f\x0a\xf9\x3f\xee\x62\xcd\x3e"
+	"\x5e\xbd\xb4\x8b\xf5\xea\xa0\x74"
+	"\x58\x68\x16\x01\x68\x4c\x31\x17"
+	"\x68\xee\x31\x40\x21\x96\xb3\x88"
+	"\x1b\x60\xb3\x51\xf1\x4d\x54\xcb"
+	"\xf3\x0b\xb1\x84\xc0\xd0\x9f\xe2"
+	"\x3b\x7f\xf4\x96\x78\xa4\x77\x43"
+	"\x4b\x96\xf7\xf3\xcc\xfc\x89\xdc"
+	"\x45\x07\xf6\xde\x32\x20\xbc\x14"
+	"\x3d\x3a\x20\xb7\xa9\xad\x7e\x21"
+	"\xc0\xeb\xe2\x4d\x2b\x88\x71\x9f"
+	"\x7e\x4d\x4b\x47\xa9\x59\xb2\x97"
+	"\xa0\x34\x9d\x30\x88\x43\x91\xdd"
+	"\x95\x21\x5a\x09\xc8\x1e\xe4\xf2"
+	"\xde\x79\x55\x77\xae\xa9\x5c\xb1"
+	"\x1c\x5f\xab\xca\x63\xf6\x4d\xd9"
+	"\x5e\x8d\x55\x99\x3d\x7f\x25\x1a"
+	"\x75\x97\x34\xdc\xde\x77\x20\x29"
+	"\x53\x0a\xe9\x93\xbd\x77\x24\xf9"
+	"\xd4\x94\xb9\x8d\xef\xe4\x1c\xcd"
+	"\x21\x3e\x78\xe5\xe5\x23\xbf\x14"
+	"\x28\x80\xa6\x10\x0d\x28\xed\xd1"
+	"\x46\x9e\xc2\xaa\x2a\xcf\xc3\x19"
+	"\xb1\x74\xfa\x0b\x48\x11\x6a\x94"
+	"\xe9\x5e\x0f\xca\xa1\x55\xdb\xdd"
+	"\x7f\x1e\x16\xf0\x95\xa3\x94\x66"
+	"\xf2\x4c\xbe\xbe\xb3\x18\x1e\x3f"
+	"\x01\x1c\xb4\xb1\x51\xde\xce\xab"
+	"\xa6\xc4\xe1\xa4\xea\x23\xd0\x79"
+	"\x46\x6a\x25\xad\x63\x2d\x7c\x45"
+	"\x1e\x14\xa9\xb1\xe2\x2e\x8f\x67"
+	"\x90\x62\x54\x0c\x02\x37\x55\xd3"
+	"\x61\xf4\x31\xba\x77\xea\x68\x18"
+	"\xeb\x2e\xdf\x8e\xf6\x35\x29\xfa"
+	"\xbb\x4e\xff\x6d\x79\x05\x88\x4c"
+	"\x9c\x13\xf9\x7a\x57\x5a\x24\x6c"
+	"\xaa\x8e\xbd\x53\x77\x8d\x0e\x0e"
+	"\x54\xec\xf6\xff\x8a\xc3\x0c\xb1"
+	"\x0a\x01\x5c\x60\xb1\x6c\xec\x05"
+	"\x09\x68\x74\x79\x39\x02\x71\x27"
+	"\xfa\x70\x14\xa7\x19\x04\xad\x55"
+	"\xcf\xf9\x7b\x64\xe2\xa0\x13\xe8"
+	"\x61\x55\x7d\x8e\x03\xa4\x28\x5d"
+	"\xa9\x12\xb4\xdb\xa3\x29\xba\x54"
+	"\xf4\xe0\x3d\xfd\xe1\x0f\x31\xc3",
+	.c_size = 512,
+	.siggen_sigver_test = true,
+	}
+};
+
 static const struct kpp_testvec dh_tv_template[] = {
 	{
 	.secret =
@@ -4545,6 +6104,52 @@ static const struct kpp_testvec ecdh_p384_tv_template[] = {
 	.b_public_size = 96,
 	.expected_a_public_size = 96,
 	.expected_ss_size = 48
+	}, {
+	.secret =
+#ifdef __LITTLE_ENDIAN
+	"\x02\x00" /* type */
+	"\x06\x00" /* len */
+	"\x00\x00", /* key_size */
+#else
+	"\x00\x02" /* type */
+	"\x00\x06" /* len */
+	"\x00\x00", /* key_size */
+#endif
+	.b_secret =
+#ifdef __LITTLE_ENDIAN
+	"\x02\x00" /* type */
+	"\x36\x00" /* len */
+	"\x30\x00" /* key_size */
+#else
+	"\x00\x02" /* type */
+	"\x00\x36" /* len */
+	"\x00\x30" /* key_size */
+#endif
+	"\x41\xCB\x07\x79\xB4\xBD\xB8\x5D"
+	"\x47\x84\x67\x25\xFB\xEC\x3C\x94"
+	"\x30\xFA\xB4\x6C\xC8\xDC\x50\x60"
+	"\x85\x5C\xC9\xBD\xA0\xAA\x29\x42"
+	"\xE0\x30\x83\x12\x91\x6B\x8E\xD2"
+	"\x96\x0E\x4B\xD5\x5A\x74\x48\xFC",
+	.b_public =
+	"\xE5\x58\xDB\xEF\x53\xEE\xCD\xE3"
+	"\xD3\xFC\xCF\xC1\xAE\xA0\x8A\x89"
+	"\xA9\x87\x47\x5D\x12\xFD\x95\x0D"
+	"\x83\xCF\xA4\x17\x32\xBC\x50\x9D"
+	"\x0D\x1A\xC4\x3A\x03\x36\xDE\xF9"
+	"\x6F\xDA\x41\xD0\x77\x4A\x35\x71"
+	"\xDC\xFB\xEC\x7A\xAC\xF3\x19\x64"
+	"\x72\x16\x9E\x83\x84\x30\x36\x7F"
+	"\x66\xEE\xBE\x3C\x6E\x70\xC4\x16"
+	"\xDD\x5F\x0C\x68\x75\x9D\xD1\xFF"
+	"\xF8\x3F\xA4\x01\x42\x20\x9D\xFF"
+	"\x5E\xAA\xD9\x6D\xB9\xE6\x38\x6C",
+	.secret_size = 6,
+	.b_secret_size = 54,
+	.b_public_size = 96,
+	.expected_a_public_size = 96,
+	.expected_ss_size = 48,
+	.genkey = true,
 	}
 };
 
@@ -7727,6 +9332,98 @@ static const struct hash_testvec aes_cmac128_tv_template[] = {
 				  "\x69\x6a\x2c\x05\x6c\x31\x54\x10",
 		.psize		= 64,
 		.ksize		= 32,
+	}, {	/* keyLen = 24 */
+		.key		= "\xb2\x03\xc6\xb8\xdd\xb2\xb3\x9b"
+				  "\xab\xe6\xed\x55\xc7\x08\x33\x7b"
+				  "\x84\x11\xa6\xcb\x63\x36\x82\xc1",
+		.plaintext	= "\xf4\x1d\x33\x7a\x3f\x74\x0f\x71"
+				  "\xdb\xc7\xfb\x18\xbe\x0e\x5c\x9e"
+				  "\x6f\xdd\x32\xbe\xbb\xd5\x33\x8e"
+				  "\x5e\x5c\xfd\xf6\x67\x8c\x4f\xc3"
+				  "\x10\x99\xf6\xc1\x04\x9d\x69\x65"
+				  "\x64\x39\x5a\xdc\x95\x67\x04\x2f"
+				  "\xfa\x50\xbe\x9d\x2c\xb2\x8f\xea"
+				  "\xde\x37\x18\xa8\xda\x73\xab\x35"
+				  "\xa3\x90\x80\xc2\x3e\x57\xef\x2d"
+				  "\x7b\x5f\x71\x4a\xc0\x12\xc6\x76"
+				  "\x04\xd8\x6a\x12\xc2\xf3\x53\x90"
+				  "\x5b\x34\x18\xf3\xcb\xe5\xc4\xd5"
+				  "\x84\x1a\xef\x5b\xdd\x09\x07\x64"
+				  "\x63\x80\x91\x80\x43\xeb\x1f\xf7"
+				  "\x2b\xe4\x8d\x9f\x49\x2f\x4f\x93"
+				  "\xec\x5d\xcb\x45\xbb\x52\xae\x0d"
+				  "\xdf\xb4\x36\xdf\x0a\xf2\x5a\x92"
+				  "\x77\xb3\x41\x11\x86\x3f\xee\x27"
+				  "\x7a\xd4\x46\x7a\xf2\xa4\x8a\xba"
+				  "\xf9\xf9\x4d\x37\x9b\x7e\x18\xa5"
+				  "\x92\x18\xbf\x31\x12\x99\xcb\x71"
+				  "\x5e\x7f\x44\xd4\xe9\x48\x8c\x39"
+				  "\x95\x67\xc4\xd4\xf7\x41\x25\x9f"
+				  "\xe6\x48\x78\xd5\x63\xb9\x12\xbb"
+				  "\x88\x2b\x5a\x13\x79\x83\x8e\x3e"
+				  "\xa8\x3d\xbc\x18\x44\x70\x03\x71"
+				  "\xcf\xb1\xf1\xae\x3a\xa1\xd0\xed"
+				  "\x2b\x39\xb5\xa1\xd7\x74\xae\x1b"
+				  "\x98\x88\x45\xed\xba\x5b\xeb\x86"
+				  "\x15\xff\x75\xc8\x2f\x51\xd9\x9b"
+				  "\xe9\x50\x4a\x12\xd4\xaa\x43\xcd"
+				  "\x18\x14\x6e\x65\x09\xa3\x3d\x1e"
+				  "\x05\xeb\x69\x1d\x0b\xed\xc1\x19"
+				  "\xc4\xc5\x58\x7e\x6e\xe9\x02\x9a"
+				  "\x9e\xbf\x44\x15\x93\xb4\x06\x01"
+				  "\xa5\x50\xda\x7c\x74\xaf\x4c\x6d"
+				  "\x7d\xe4\xb1\x31\xec\x77\x5a\x2e"
+				  "\x66\xdd\xc7\xdb\xac\xe8\x1e\xf3"
+				  "\x91\xe3\x14\x56\x11\xbb\x03\xb7"
+				  "\x1c\x55\x95\x66\x1b\x28\xf7\x8b"
+				  "\xa1\x04\x67\x8c\xdd\x93\x7d\xb3"
+				  "\x15\x55\xd1\x97\x6c\x23\x57\x3a"
+				  "\xb6\xf6\xa4\x6b\x3a\xed\x72\xe4"
+				  "\x71\x8f\xc4\x69\x2b\x6f\x84\x37"
+				  "\x74\x33\xff\x03\x23\xd5\xe9\x35"
+				  "\x4f\x4f\x43\x84\x9f\x75\x46\xe1"
+				  "\xbf\xc0\xd5\xa0\x5f\x6b\xbe\xa0"
+				  "\x08\x8a\x62\xc6\xc9\xcc\x02\x59"
+				  "\xd9\x70\x0c\xb2\x05\x17\xfa\x8c"
+				  "\x96\x03\xd7\xf8\x8e\x7b\x6e\xd3"
+				  "\x55\x8d\x36\x2e\xc8\x5e\xa7\x81"
+				  "\xe6\x05\x1b\x1e\x61\x7f\x39\x27"
+				  "\x48\xd2\xff\x0e\x49\x9c\x6f\xf6"
+				  "\xee\x9e\xfe\x07\xec\xa4\xa6\x23"
+				  "\xa8\xc3\x46\x02\xf3\x4b\x2b\xd9"
+				  "\x7e\x55\x4f\x75\x0d\xe9\x2a\x94"
+				  "\x0a\xc4\x9d\x35\x9f\x17\x97\x9c"
+				  "\x90\x2a\x79\xf4\x38\x26\xd6\x35"
+				  "\x7b\x28\x07\x8c\x14\x44\xeb\xe5"
+				  "\x07\x9f\x26\xe0\x38\x00\x20\x67"
+				  "\x92\x36\x2d\xc3\xd9\x21\x4c\xb0"
+				  "\x50\x5a\x46\xdf\x30\x33\x79\xa5"
+				  "\x36\x87\x75\x2e\x50\xbd\xcb\x38"
+				  "\x88\x77\x56\xeb\xc8\xe8\xf7\xd6"
+				  "\x0a\xa9\x27\xb1\xcb\x0c\xbd\xf0"
+				  "\x56\x58\x49\xe0\x7c\x73\x5c\x5d"
+				  "\x8b\xee\xc9\x2f\xd6\x2b\x11\xb9"
+				  "\xe7\xf9\xcf\x96\x12\xa2\x08\x99"
+				  "\x3a\xc2\x53\x8d\x90\x04\x93\xff"
+				  "\xb7\x4c\x59\x61\x76\xb3\x31\x57"
+				  "\x0d\x68\x8e\xd0\x82\x31\x99\x5e"
+				  "\x37\xb8\xb5\x5f\x0b\x74\x3d\xaf"
+				  "\xb9\xbf\x93\xae\xcb\x6a\xc6\x1a"
+				  "\xbf\x4b\x61\x70\x89\x48\x71\x9d"
+				  "\x58\xdb\x0b\x4f\x52\xc2\x59\xce"
+				  "\x70\x1f\xf8\x57\x1e\x52\x4a\x18"
+				  "\xbd\xe2\xa9\xe8\x50\x89\x33\x19"
+				  "\xb8\xe1\x93\xdb\x00\x0a\x7d\x2f"
+				  "\x0f\xd4\xb1\x56\x8d\xfa\x92\x63"
+				  "\xa8\xe4\x38\xfe\x84\x7f\xfb\x52"
+				  "\x27\x61\x7b\x06\x77\x81\x71\x09"
+				  "\x31\x98\x4a\xe4\x25\x01\xb2\xa8"
+				  "\x1d\x7c\xd6\x19\xa4\x71\x7a\x38"
+				  "\xe9\x21\x69\xc2\x93\x64",
+		.digest		= "\x65\x65\xed\xf1\xf7\xc8\x26\xc2"
+				  "\xa4\x95\x9a\x9d\x48\xdf\xd0\x03",
+		.psize		= 670,
+		.ksize		= 24,
 	}
 };
 
@@ -24137,6 +25834,40 @@ static const struct drbg_testvec drbg_pr_hmac_sha256_tv_template[] = {
 	},
 };
 
+static const struct drbg_testvec drbg_pr_ctr_aes192_tv_template[] = {
+	{
+		.entropy = (unsigned char *)
+			"\x96\x5f\xa6\xa2\xc4\x00\x9c\x04\x5c\xff"
+			"\x72\x8f\x24\x42\x38\x26\x25\x08\xbf\x00"
+			"\x8d\xcc\xef\x30\x1d\x22\x59\x1b\x09\x78"
+			"\x73\xd1\x24\xc3\xf8\xc4\x08\xf8\x91\x61",
+		.entropylen = 40,
+		.entpra = (unsigned char *)
+			"\x97\x18\xb8\x80\x42\xa4\xb6\x07\x67\xca"
+			"\xe0\xff\xc7\x97\xe6\x88\xb0\xfe\x38\x19"
+			"\xdb\x2a\x4b\xcb",
+		.entprb = (unsigned char *)
+			"\x73\xe3\xf0\xfd\xf3\x80\xe5\xe6\x12\x6f"
+			"\xf9\xf3\xad\xb3\x4f\xf5\xbc\xc4\x5f\xba"
+			"\x0c\xe1\xd2\x48",
+		.entprlen = 24,
+		.expected = (unsigned char *)
+			"\x90\x9f\xd2\xc9\x71\x92\x1c\x77\x32\x2e"
+			"\x1d\xf2\x73\xa0\x7d\x19\xf5\x27\x13\x3e"
+			"\x56\xa9\xbe\x38\x8c\x29\x8c\xa6\x83\x08"
+			"\x73\xd1\xe8\x47\xb8\xcc\x65\x4d\xd3\xc5"
+			"\x7c\x1c\x8b\x25\xfd\x39\x09\xbe\x84\x7c"
+			"\x1a\x28\x11\x08\x94\x0f\xe7\xf0\x6d\x9d"
+			"\xb6\x0a\xd7\xb4",
+		.expectedlen = 64,
+		.addtla = NULL,
+		.addtlb = NULL,
+		.addtllen = 0,
+		.pers = NULL,
+		.perslen = 0,
+	}
+};
+
 static const struct drbg_testvec drbg_pr_ctr_aes128_tv_template[] = {
 	{
 		.entropy = (unsigned char *)
@@ -24253,6 +25984,185 @@ static const struct drbg_testvec drbg_pr_ctr_aes128_tv_template[] = {
 	},
 };
 
+static const struct drbg_testvec drbg_pr_ctr_aes256_tv_template[] = {
+	{
+		.entropy = (unsigned char *)
+			"\x53\x43\x46\xa3\xe0\xba\xa6\x5d\x7a\x51\x87\x1b\x6d"
+			"\x63\x3a\x6f\x1e\xfa\x9f\xf5\x5d\xfd\xe3\x21\x2c\x95"
+			"\x02\x9a\xdf\x23\x87\xd9\x0c\xbe\x99\x82\x15\x09\x97"
+			"\x5d\x82\x4f\xd8\x26\xc4\x7d\x2a\xbc",
+		.entropylen = 48,
+		.entpra = (unsigned char *)
+			"\x28\x60\x91\x70\x9d\x2e\x91\xbf\xef\x66\x54\xb8\x44"
+			"\x9d\x5e\x55\x64\xb6\xc2\xfe\x1f\x3e\x4e\x6e\x59\xfe"
+			"\x77\x64\x61\xbf\xb0\xdc",
+		.entprb = (unsigned char *)
+			"\xe6\x8e\x12\x7b\x2e\x5b\x9d\x65\x39\xc2\x6f\x7e\x78"
+			"\xbe\xc1\x33\x14\xe3\x7f\xeb\xdb\x71\x05\xd1\xd6\x5f"
+			"\x0b\xde\x23\xc8\x7d\x38",
+		.entprlen = 32,
+		.expected = (unsigned char *)
+			"\xda\x63\x89\xd1\x51\xe3\xb2\xb3\x32\xfb\xcc\x02\xb2"
+			"\xcc\x5a\xf4\xf6\x98\x35\xe9\xfb\xbe\x19\x13\x0f\x91"
+			"\xcb\xf3\x0a\x3c\x24\x58\x0e\xbf\xb4\x31\x1b\x34\x87"
+			"\xe9\xd6\xa6\x1d\xe9\xdc\x28\x42\xc1\x07\x15\x21\x58"
+			"\xb0\xe7\x5a\x93\x2c\x38\xe5\xb8\xdd\xff\x10\xe6",
+		.expectedlen = 64,
+		.addtla = (unsigned char *)
+			"\xf4\x10\x40\xef\xf2\xa7\x01\x02\x87\xc2\xa7\x6d\x78"
+			"\x67\xa6\x6e\x61\x29\x9b\xe1\xe1\x24\x7a\x6b\x22\xa6"
+			"\x4b\x82\x9a\xda\x86\x24",
+		.addtlb = (unsigned char *)
+			"\x94\xef\xe6\x8e\x21\x85\x64\x6a\x17\xbb\x8a\x83\x26"
+			"\x81\x42\xa7\xb7\xa0\x27\x86\xc1\xee\x90\xab\x8f\x73"
+			"\x99\xcc\xf5\x43\xde\x26",
+		.addtllen = 32,
+		.pers = (unsigned char *)
+			"\x02\x0d\x73\x40\x56\x9e\x68\xd9\x92\xe6\xe8\xdd\xfb"
+			"\xea\xd9\xf9\x93\xc7\x4d\x97\x1e\x43\x39\xfe\x91\x07"
+			"\x4f\x87\xe9\xd7\xd7\x77",
+		.perslen = 32,
+	},
+};
+
+static const struct drbg_testvec drbg_pr_sha1_tv_template[] = {
+	{
+		.entropy = (unsigned char *)
+			"\x21\x29\x56\x39\x07\x83\x38\x1d\xbf\xc6"
+			"\x36\x2d\xd0\xda\x9a\x09\x52\x80\x98\x7f"
+			"\xc5\xe2\x7a\x49",
+		.entropylen = 24,
+		.entpra = (unsigned char *)
+			"\x2e\xdb\x39\x6e\xeb\x89\x60\xf7\x79\x43"
+			"\xc2\xa5\x90\x75\xa7\x86",
+		.entprb = (unsigned char *)
+			"\x30\xb5\x65\xb6\x3a\x50\x12\x67\x69\x40"
+			"\xd3\xef\x17\xd9\xe9\x96",
+		.entprlen = 16,
+		.expected = (unsigned char *)
+			"\xca\x50\xec\x95\xc7\xc3\x8a\x58\x12\x9f"
+			"\xd3\x75\x23\xd1\xf2\x59\x8c\xe6\xde\x98"
+			"\xa6\xf1\x07\x72\x4c\x55\x46\xbe\xad\xaa"
+			"\x5b\xaf\xbf\x1e\x62\xd2\x74\x84\x3f\x11"
+			"\x07\xc3\x45\xb1\x28\x81\x63\xf1\xdd\x24"
+			"\xaf\xb2\x7a\x63\xd3\x9e\xe2\xaa\x77\x0d"
+			"\xfc\xa6\x68\xeb\x13\x4a\xbe\x08\x05\x78"
+			"\xb2\xf8\xd1\xa3\xa8\x99\xf5\xa0\x0a\xf7",
+		.expectedlen = 80,
+		.addtla = NULL,
+		.addtlb = NULL,
+		.addtllen = 0,
+		.pers = NULL,
+		.perslen = 0,
+	},
+};
+
+static const struct drbg_testvec drbg_pr_sha384_tv_template[] = {
+	{
+		.entropy = (unsigned char *)
+			"\x83\x6e\xa2\x58\xe9\x7d\xb5\x14\xdc\xfa"
+			"\xcc\xc2\x4d\xfe\x63\x70\x80\x27\xa3\x30"
+			"\x74\xe8\xd4\x14\x43\xb7\x45\x70\x3b\x34"
+			"\x5c\x81\x0f\x04\x03\x58\x44\xf6\xaa\xe9"
+			"\x44\x44\x86\x51\x1d\x12\x9f\x7c",
+		.entropylen = 48,
+		.entpra = (unsigned char *)
+			"\x84\x39\x5a\x07\x71\x30\x93\x91\x6c\xf9"
+			"\xc2\x46\x3a\xb9\xb9\x98\x31\x27\x3b\xcd"
+			"\x5d\x21\xf3\x0c\x1a\xe4\x7a\xa6\x21\x8f"
+			"\x7c\xa2",
+		.entprb = (unsigned char *)
+			"\xed\x21\x83\x9b\xfe\x80\x27\x47\x88\x72"
+			"\xd0\x4b\x68\xa2\x76\xd8\x41\xfd\xd8\x7c"
+			"\xbf\x58\xa3\xb7\x7f\xd8\xba\x3d\x28\x53"
+			"\x94\x0c",
+		.entprlen = 32,
+		.expected = (unsigned char *)
+			"\x87\xa9\x18\xcb\x5e\xb0\xc5\xa0\xae\x3e"
+			"\xd3\x03\x1d\x2a\x39\x90\x4a\xf0\xb3\x38"
+			"\x14\xb7\xf3\xe2\x18\x8e\x05\x7c\xb2\xb2"
+			"\xc4\x41\x44\xa5\xfa\xcb\x09\x4c\x33\x4e"
+			"\x6e\xc7\x0d\x43\x66\xf8\x61\x5c\xcf\x98"
+			"\xcd\xf3\x39\x2d\x0d\x64\x41\x59\xf7\x21"
+			"\xda\xf2\xa7\x7a\xd7\xdb\xec\x72\xde\x6a"
+			"\x59\x36\x0e\xcd\x4f\xad\x60\x15\xe5\x91"
+			"\x70\xad\xb9\x6a\xda\x2c\xaf\x39\x51\xdf"
+			"\x3c\x6f\xd8\x93\x6c\x7d\x98\x62\xd9\xb8"
+			"\x1c\x84\xa9\xa1\xe6\xdf\x42\x3d\x74\x3d"
+			"\x15\x7e\xe5\x0a\xa8\xb9\x2f\xed\xd4\x20"
+			"\x52\xce\x4f\x17\xcf\x62\x9e\xae\x7d\xc2"
+			"\xa2\xd9\xf4\x8d\x73\x73\xf3\x0c\xe5\x3d"
+			"\xfe\x2e\x2e\x5f\xeb\xe9\x71\x8a\x87\x87"
+			"\xd3\xe1\xf3\x94\x1d\x0e\x9c\x47\x23\x33"
+			"\x25\x45\xd1\xa9\x9e\x94\x6c\xf5\xe3\xb3"
+			"\x35\xf9\x54\x91\x22\x69\x82\x89\x1d\xe7"
+			"\x5f\x31\x67\xd8\x80\xca\xda\x8a\x1e\x76"
+			"\xb2\x0e",
+		.expectedlen = 192,
+		.addtla = NULL,
+		.addtlb = NULL,
+		.addtllen = 0,
+		.pers = NULL,
+		.perslen = 0,
+	},
+};
+
+static const struct drbg_testvec drbg_pr_sha512_tv_template[] = {
+	{
+		.entropy = (unsigned char *)
+			"\x73\xc9\xb1\x15\xb7\xef\xb0\xa6\x32\x44"
+			"\xd7\x49\x3a\xe5\x82\x05\x99\xd7\xce\xe5"
+			"\xca\x05\x4d\xb2\xf7\x26\x9b\xa7\xf6\x21"
+			"\xbd\xca\xc2\x04\xe6\xde\x78\x9b\x03\x94"
+			"\xfb\xbe\x66\x63\x46\x6e\xfc\xea",
+		.entropylen = 48,
+		.entpra = (unsigned char *)
+			"\xcf\xce\xf3\x77\x6b\x37\x64\x9a\x7f\x6d"
+			"\x2b\x48\xf4\x43\xda\x79\xa2\xf2\xf8\x1d"
+			"\x04\xf3\xaf\x98\x53\xa9\xe6\x96\xc4\x48"
+			"\x74\x40",
+		.entprb = (unsigned char *)
+			"\xd0\x63\x8e\x28\xca\xe8\xd1\xc0\xf5\x72"
+			"\x09\xd6\x77\xd8\x89\xd1\x95\xa6\x72\x02"
+			"\x3c\xb8\xad\xe3\x9f\x79\x49\x89\xe1\xda"
+			"\xee\x34",
+		.entprlen = 32,
+		.expected = (unsigned char *)
+			"\x04\x74\x4d\x1d\x42\x60\x19\x95\xfa\x3b"
+			"\x10\x1d\xed\x3d\x25\x31\xcb\xf4\x5a\xfd"
+			"\x83\x12\x0d\x58\xeb\x26\x59\x4a\x86\x3b"
+			"\xd8\x31\x83\x11\xb0\x8d\x3d\xf4\xc5\x71"
+			"\xa9\xc2\x6d\xff\x63\xa3\xe9\x91\x3a\x9a"
+			"\x17\xa7\xc4\x55\x18\x6f\xdf\xdd\x90\xc6"
+			"\x64\xa8\x4b\x73\xa1\x10\x6a\x5a\x82\xf7"
+			"\x41\xbd\x4c\x7a\x48\xbd\x04\x6c\x26\x8d"
+			"\x89\x19\xef\xc9\x41\xf8\xb4\x5a\x3c\x3d"
+			"\x89\xcf\x37\x14\x1b\x5c\x41\xb1\x0f\xf5"
+			"\x43\xa6\x92\x62\x72\xd6\x23\xad\x8e\xcc"
+			"\xd0\x26\x55\x20\x90\xad\xcf\xac\xb1\x24"
+			"\xf4\x7c\x4a\xd6\x2b\xe9\x0e\xa5\xa0\xa7"
+			"\x08\x7d\x81\x45\x84\x45\x81\x3a\xf8\x8f"
+			"\xfb\x5a\x8c\x35\x19\xf9\x77\x13\x1c\xc8"
+			"\x51\xcb\x44\x54\xb0\xa7\x56\xc8\x37\x3f"
+			"\x05\x23\x82\x43\x5a\xb9\x34\x71\x8c\x95"
+			"\x51\x77\x36\x33\x89\xc0\x6b\x0b\x50\x73"
+			"\x47\x8e\x84\xd2\x53\xff\x02\xa3\xf1\xbe"
+			"\xf1\xbb\xf1\x33\x8f\x77\xf9\x2f\x02\x9f"
+			"\x63\x8a\x46\x91\xc4\x8c\x47\x0d\x30\xd2"
+			"\x30\xf0\x07\xf5\x45\xe0\x22\xf6\x6c\x78"
+			"\xa1\x30\x69\x78\x14\xaa\x55\xd2\x00\x0a"
+			"\x49\x55\x3b\xef\x35\xfa\xb5\x80\x8e\x2f"
+			"\x3c\xbb\x38\xc4\x05\x61\x1f\xa8\x14\x44"
+			"\x12\x4e\x3f\x89\xe1\xe8",
+		.expectedlen = 256,
+		.addtla = NULL,
+		.addtlb = NULL,
+		.addtllen = 0,
+		.pers = NULL,
+		.perslen = 0,
+	},
+};
+
 /*
  * SP800-90A DRBG Test vectors from
  * http://csrc.nist.gov/groups/STM/cavp/documents/drbg/drbgtestvectors.zip
@@ -24383,6 +26293,45 @@ static const struct drbg_testvec drbg_nopr_sha256_tv_template[] = {
 	},
 };
 
+static const struct drbg_testvec drbg_nopr_hmac_sha384_tv_template[] = {
+	{
+		.entropy = (unsigned char *)
+			"\xa1\xdc\x2d\xfe\xda\x4f\x3a\x11\x24\xe0"
+			"\xe7\x5e\xbf\xbe\x5f\x98\xca\xc1\x10\x18"
+			"\x22\x1d\xda\x3f\xdc\xf8\xf9\x12\x5d\x68"
+			"\x44\x7a\xba\xe5\xea\x27\x16\x65\x40\x51"
+			"\x52\x68\xa4\x93\xa9\x6b\x51\x87",
+		.entropylen = 48,
+		.expected = (unsigned char *)
+			"\x22\x82\x93\xe5\x9b\x1e\x45\x45\xa4\xff"
+			"\x9f\x23\x26\x16\xfc\x51\x08\xa1\x12\x8d"
+			"\xeb\xd0\xf7\xc2\x0a\xce\x83\x7c\xa1\x05"
+			"\xcb\xf2\x4c\x0d\xac\x1f\x98\x47\xda\xfd"
+			"\x0d\x05\x00\x72\x1f\xfa\xd3\xc6\x84\xa9"
+			"\x92\xd1\x10\xa5\x49\xa2\x64\xd1\x4a\x89"
+			"\x11\xc5\x0b\xe8\xcd\x6a\x7e\x8f\xac\x78"
+			"\x3a\xd9\x5b\x24\xf6\x4f\xd8\xcc\x4c\x8b"
+			"\x64\x9e\xac\x2b\x15\xb3\x63\xe3\x0d\xf7"
+			"\x95\x41\xa6\xb8\xa1\xca\xac\x23\x89\x49"
+			"\xb4\x66\x43\x69\x4c\x85\xe1\xd5\xfc\xbc"
+			"\xd9\xaa\xae\x62\x60\xac\xee\x66\x0b\x8a"
+			"\x79\xbe\xa4\x8e\x07\x9c\xeb\x6a\x5e\xaf"
+			"\x49\x93\xa8\x2c\x3f\x1b\x75\x8d\x7c\x53"
+			"\xe3\x09\x4e\xea\xc6\x3d\xc2\x55\xbe\x6d"
+			"\xcd\xcc\x2b\x51\xe5\xca\x45\xd2\xb2\x06"
+			"\x84\xa5\xa8\xfa\x58\x06\xb9\x6f\x84\x61"
+			"\xeb\xf5\x1b\xc5\x15\xa7\xdd\x8c\x54\x75"
+			"\xc0\xe7\x0f\x2f\xd0\xfa\xf7\x86\x9a\x99"
+			"\xab\x6c",
+		.expectedlen = 192,
+		.addtla = NULL,
+		.addtlb = NULL,
+		.addtllen = 0,
+		.pers = NULL,
+		.perslen = 0,
+	},
+};
+
 static const struct drbg_testvec drbg_nopr_hmac_sha256_tv_template[] = {
 	{
 		.entropy = (unsigned char *)
@@ -25128,6 +27077,283 @@ static const struct cipher_testvec cast5_ctr_tv_template[] = {
 	},
 };
 
+static const struct drbg_testvec drbg_pr_hmac_sha1_tv_template[] = {
+	{
+		.entropy = (unsigned char *)
+					"\x68\x0f\xac\xe9\x0d\x7b\xca\x21\xd4\xa0"
+					"\xed\xb7\x79\x9e\xe5\xd8\xb7\xbe\x9e\xed"
+					"\xdd\x0e\x3b\x4b",
+		.entropylen = 24,
+		.entpra = (unsigned char *)
+					 "\x7c\xaf\xe2\x31\x63\x0a\xa9\x5a\x74\x2c"
+					 "\x4e\x5f\x5f\x22\xc6\xa4",
+		.entprb = (unsigned char *)
+					 "\x1c\x0d\x77\x92\x89\x88\x27\x94\x8a\x58"
+					 "\x9f\x82\x2d\x1a\xf7\xa6",
+		.entprlen = 16,
+		.expected = (unsigned char *)
+						"\x68\x00\x4b\x3a\x28\xf7\xf0\x1c\xf9\xe9"
+						"\xb5\x71\x20\x79\xef\x80\x87\x1b\x08\xb9"
+						"\xa9\x1b\xcd\x2b\x9f\x09\x4d\xa4\x84\x80"
+						"\xb3\x4c\xaf\xd5\x59\x6b\x0c\x0a\x48\xe1"
+						"\x48\xda\xbc\x6f\x77\xb8\xff\xaf\x18\x70"
+						"\x28\xe1\x04\x13\x7a\x4f\xeb\x1c\x72\xb0"
+						"\xc4\x4f\xe8\xb1\xaf\xab\xa5\xbc\xfd\x86"
+						"\x67\xf2\xf5\x5b\x46\x06\x63\x2e\x3c\xbc",
+		.expectedlen = 80,
+		.addtla = (unsigned char *)
+					 "\xdc\x36\x63\xf0\x62\x78\x9c\xd1\x5c\xbb"
+					 "\x20\xc3\xc1\x8c\xd9\xd7",
+		.addtlb = (unsigned char *)
+					 "\xfe\x85\xb0\xab\x14\xc6\x96\xe6\x9c\x24"
+					 "\xe7\xb5\xa1\x37\x12\x0c",
+		.addtllen = 16,
+		.pers = (unsigned char *)
+				  "\xf5\x8c\x40\xae\x70\xf7\xa5\x56\x48\xa9"
+				  "\x31\xa0\xa9\x31\x3d\xd7",
+		.perslen = 16,
+	},
+};
+
+static const struct drbg_testvec drbg_pr_hmac_sha384_tv_template[] = {
+	{
+		.entropy = (unsigned char *)
+					  "\x8b\x28\x5c\xe6\xb4\xda\x70\xc8\x3f\xd7"
+					  "\x2a\xab\x1b\x4b\xe6\x21\x01\xbf\x9b\x29"
+					  "\xe1\x68\x72\x6e\xa2\xf6\x70\xaa\xb0\xde"
+					  "\xae\xfc\x5d\xa3\x40\x4c\x49\x4c\x60\x19"
+					  "\xea\x33\x67\x9e\x37\xce\xc3\x08",
+		.entropylen = 48,
+		.entpra = (unsigned char *)
+					 "\xda\xb1\x3e\x0c\xb0\x60\xf6\x6c\x1c\x83"
+					 "\xfc\x6f\xba\x46\x47\x7d\x1a\x3c\x80\x2e"
+					 "\xdd\x75\x94\xdb\x0b\x29\x7d\xed\xb9\xcc"
+					 "\xbc\x80",
+		.entprb = (unsigned char *)
+					 "\x0c\x81\x7f\x05\x65\x8f\xb9\xb4\xc9\x99"
+					 "\x38\xae\x21\x40\x16\x0c\x4a\x16\xd5\x48"
+					 "\x63\x4a\x35\x3b\xc2\x85\xcb\x38\xd0\xe9"
+					 "\x32\x43",
+		.entprlen = 32,
+		.expected = (unsigned char *)
+						"\x51\xe0\x42\xdd\x56\xa1\x93\x90\x8c\x90"
+						"\x18\xc2\x5f\x1c\x1a\x8b\x5e\x27\x34\xb0"
+						"\x55\xc3\xb7\xfd\xe6\xa8\xba\x9e\xc2\xb9"
+						"\x59\x34\x9d\xf2\x92\x95\xab\xb0\xa2\x4b"
+						"\x47\x15\xf9\x8d\x31\xde\x0a\x36\x9e\x62"
+						"\x62\xc2\xb2\xcd\x49\xc5\x46\x2b\x7a\xe2"
+						"\x84\xe9\x21\xf5\xad\x2e\xc0\x13\xed\xc1"
+						"\x61\x13\x43\xc2\x28\x68\x3f\x41\x70\xf3"
+						"\x4a\x75\x85\x4b\x1b\x65\x6d\x22\x6e\x29"
+						"\x41\x72\xd4\x88\xc1\x0a\x41\x5f\x09\xde"
+						"\xe7\x09\x84\xb9\xc4\x9e\x8d\x36\x86\x31"
+						"\x92\x30\x1d\x17\x62\x14\x5e\x0d\x9e\x94"
+						"\xe9\x9b\xd3\x0c\xe8\x49\x04\x38\xed\x05"
+						"\x0f\x41\x8c\xf4\xba\x0b\x07\xfe\x90\xa8"
+						"\x2d\x1c\xcf\x38\x57\x8d\x99\xed\xf0\x51"
+						"\x8c\x4a\x75\x8a\x19\x9d\xb4\xd3\x53\x3c"
+						"\x4d\xbc\x55\xb1\xda\x19\x84\x0b\x8f\x36"
+						"\x5a\x57\x4a\xa0\x16\x47\x81\x90\x32\xdc"
+						"\x0a\xd6\x41\x38\x8c\x20\x93\xeb\xd4\xab"
+						"\x5d\x99",
+		.expectedlen = 192,
+		.addtla = (unsigned char *)
+					 "\x88\x1a\x18\x74\xc8\x00\xdb\x06\x8b\x59"
+					 "\x13\xd1\x95\x05\x8d\x07\x26\x45\x8d\xe3"
+					 "\x78\x2f\xf5\x30\xaf\x1a\x76\x1f\x96\x28"
+					 "\x54\x7f",
+		.addtlb = (unsigned char *)
+					 "\x0c\x27\xcf\x27\x1b\xd7\x93\x1d\x18\x7e"
+					 "\xc6\xf5\x60\x38\x51\x96\x74\x46\x8f\xa2"
+					 "\xe7\xe6\xf9\x94\x90\x4c\x9f\x1a\xfa\x34"
+					 "\x69\x39",
+		.addtllen = 32,
+		.pers = (unsigned char *)
+				  "\x72\x3c\x0f\x28\x7d\xb4\xaf\x28\x5c\x19"
+				  "\x5c\xeb\xb1\x10\x4a\x10\x6f\x22\xe8\xb2"
+				  "\x43\xfd\xcd\x05\x66\x22\x8a\xb5\xf2\x27"
+				  "\xa9\xe3",
+		.perslen = 32,
+	},
+};
+
+static const struct drbg_testvec drbg_pr_hmac_sha512_tv_template[] = {
+	{
+		.entropy = (unsigned char *)
+					  "\x3a\xca\x6b\x55\x56\x15\x21\x00\x7c\x9e"
+					  "\xce\x08\x5e\x9a\x66\x35\xe3\x46\xfa\x80"
+					  "\x43\x35\xd6\xad\x42\xeb\xd6\x81\x4c\x01"
+					  "\x7f\xa8\xaa\x7f\xd3\xc3\xdd\x5d\x03\xd9"
+					  "\xb8\xef\xc7\xf7\x05\x74\x58\x1f",
+		.entropylen = 48,
+		.entpra = (unsigned char *)
+					 "\x4c\xc1\x9f\xae\x5a\x45\x6f\x8a\x53\xa6"
+					 "\x56\xd2\x3a\x0b\x66\x5d\x6d\xdf\x7f\x43"
+					 "\x02\x0a\x5f\xeb\xbb\x55\x27\x14\xe4\x47"
+					 "\x56\x5d",
+		.entprb = (unsigned char *)
+					 "\x63\x73\x86\xb3\xab\x33\xf7\x8f\xd9\x75"
+					 "\x1c\x7b\x7e\x67\xe1\xe1\x5f\x6e\x50\xdd"
+					 "\xc5\x48\xa1\xeb\x58\x13\xf6\xd0\xd4\x83"
+					 "\x81\xbf",
+		.entprlen = 32,
+		.expected = (unsigned char *)
+						"\x54\x66\x64\x04\x2b\xef\x33\x06\x4d\xa2"
+						"\x8a\x57\x18\xf2\xc2\xe5\xf7\x2d\x77\x25"
+						"\xe3\xfb\xe8\x7a\xd2\xee\x90\xfb\xfe\x6c"
+						"\x11\x4e\xd3\x64\x40\xfb\xbc\xcf\x29\x69"
+						"\x8b\x43\x60\xbc\x4a\xd7\x46\x50\xde\x13"
+						"\x82\x58\x38\x10\x6a\xdc\x53\x00\x2b\xc3"
+						"\x89\xee\x90\x06\x91\x64\x9b\x97\x2f\x31"
+						"\x87\xb8\x4d\x05\xce\xcc\x8f\xd0\x34\x49"
+						"\x7d\xd9\x9c\x6c\x99\x7d\x19\x14\xb4\xef"
+						"\x83\x8d\x84\xab\xf2\x3f\xae\x7f\x3a\xc9"
+						"\xef\xdc\xdc\x04\xc0\x03\xac\x64\x2c\x51"
+						"\x26\xb0\x0f\x9f\x24\xbf\x14\x31\xa4\xf1"
+						"\x9e\xf0\xb5\xf3\xd2\x30\xaa\xb3\xfd\xf0"
+						"\x91\xba\x31\xb7\xdd\xca\xcd\xf2\x56\x6f"
+						"\x2c\xfa\xb3\x0f\x55\xb3\x12\x3e\x73\x38"
+						"\x29\xb6\x97\xb7\xc8\xb2\x48\x42\x0a\xb9"
+						"\x8b\xa6\xf1\x1b\x01\x71\x75\x25\x63\x68"
+						"\xe8\xd8\x36\x11\x02\xc9\xe6\xd5\x73\x86"
+						"\xbe\xcb\xea\xbd\xa0\x92\xdd\x57\xae\xc6"
+						"\x5b\xc2\x0e\xbe\xe7\x8e\xea\x72\x94\x57"
+						"\x1e\x16\x8c\x45\x40\x66\xd2\x56\xb8\x1b"
+						"\xb8\xb7\xbb\x46\x92\x07\xa1\x8e\xbe\xdb"
+						"\xb4\x34\x8f\xbe\x97\xa4\xd8\x6d\x2b\xd0"
+						"\x95\xc4\x1f\x6d\xe5\x9a\xa0\x80\x0e\x13"
+						"\x1e\x98\x18\x18\x86\xa2\x63\x3c\xdc\xc5"
+						"\x50\x91\x4d\x83\xb3\x27",
+		.expectedlen = 256,
+		.addtla = (unsigned char *)
+					 "\xb3\x9c\x43\x53\x9f\xdc\x24\x34\x30\x85"
+					 "\xcb\xb6\x5b\x8d\x36\xc5\x47\x32\x47\x6d"
+					 "\x78\x11\x04\xc3\x55\xc3\x91\xa9\x51\x31"
+					 "\x3a\x30",
+		.addtlb = (unsigned char *)
+					 "\xb6\x85\x0e\xdd\x46\x22\x67\x5e\xf5\xa5"
+					 "\x07\xea\xb9\x11\xe2\x49\xd6\x3f\xcf\x62"
+					 "\xf3\x30\xcc\x8a\x16\xbb\x2c\xcc\x58\x58"
+					 "\xde\x5d",
+		.addtllen = 32,
+		.pers = (unsigned char *)
+				  "\x4b\xc9\xa4\x85\xec\x84\x0d\x37\x7a\xe4"
+				  "\x50\x4a\xa1\xdf\x41\xe4\x44\xc4\x23\x16"
+				  "\x87\xf3\xd7\x85\x1c\x26\xc2\x75\xbc\x68"
+				  "\x74\x63",
+		.perslen = 32,
+	},
+};
+
+static const struct drbg_testvec drbg_nopr_sha1_tv_template[] = {
+	{
+		.entropy = (unsigned char *)
+			"\x13\x6c\xf1\xc1\x74\xe5\xa0\x9f\x66\xb9"
+			"\x62\xd9\x94\x39\x65\x25\xff\xf1\xc6\x64"
+			"\x5f\x19\x23\x1f",
+		.entropylen = 24,
+		.expected = (unsigned char *)
+			"\x0e\x28\x13\x0f\xa5\xca\x11\xed\xd3\x29"
+			"\x3c\xa2\x6f\xdb\x8a\xe1\x81\x06\x11\xf7"
+			"\x87\x15\x08\x2e\xd3\x84\x1e\x74\x86\xf1"
+			"\x66\x77\xb2\x8e\x33\xff\xe0\xb9\x3d\x98"
+			"\xba\x57\xba\x35\x8c\x13\x43\xab\x2a\x26"
+			"\xb4\xeb\x79\x40\xf5\xbc\x63\x93\x84\x64"
+			"\x1e\xe8\x0a\x25\x14\x03\x31\x07\x62\x68"
+			"\xbd\x1c\xe7\x02\xad\x53\x4d\xda\x0e\xd8",
+		.expectedlen = 80,
+		.addtla = NULL,
+		.addtlb = NULL,
+		.addtllen = 0,
+		.pers = NULL,
+		.perslen = 0,
+	},
+};
+
+static const struct drbg_testvec drbg_nopr_sha384_tv_template[] = {
+	{
+		.entropy = (unsigned char *)
+			"\x9e\xf0\xb0\x03\x81\xd6\xc8\xc5\x4d\x08"
+			"\xfc\xad\xc6\xf5\xef\x33\x11\x34\xbb\x98"
+			"\x63\x73\xf6\x5c\x6a\x14\xf5\x53\xbc\xb6"
+			"\xc5\x5d\x9f\xce\x26\xad\xa7\xb1\xde\x39"
+			"\x59\x03\x12\xbd\x9d\x81\xc4\xf5",
+		.entropylen = 48,
+		.expected = (unsigned char *)
+			"\x66\x3f\xfb\x62\x5e\x62\xc4\xeb\x67\xd7"
+			"\x17\x7a\x6a\xbb\x80\x8a\x9f\x68\xc2\xd5"
+			"\x84\x0f\x19\x99\x2c\x11\xea\x3a\x63\x5d"
+			"\x05\xb5\x37\xfa\xe1\xf1\x74\x6c\x13\x14"
+			"\xe1\xa7\x5e\x14\x1c\x2e\x09\x41\x87\xd1"
+			"\x7b\x9d\xaa\xe1\x44\x2e\x41\xd3\xa0\xd1"
+			"\xfe\xa9\x4d\x8e\xf9\xd8\x40\x11\x13\x79"
+			"\xa5\x2e\x6c\x7f\xfa\xfa\x7e\xe8\x3b\x24"
+			"\x4c\xed\x12\x96\x13\xd5\xb8\xbb\x08\x9e"
+			"\x7e\xa2\x5d\xe1\xc2\x98\x97\x73\x5c\xf9"
+			"\x56\x95\x04\x3a\x64\x8a\x2e\xf6\xfd\x4a"
+			"\xa7\x4c\xe8\x32\x8a\x55\x50\xda\x8d\xdb"
+			"\x51\xf9\x8a\xdc\xdc\x10\x8e\x45\x56\x03"
+			"\xf6\xf1\x8f\x5a\x50\x01\x6f\x3e\x8e\xbc"
+			"\xb2\x44\xa1\x6b\xc6\xb6\xe5\x54\xa7\x54"
+			"\x61\x53\xc1\x2f\x52\x2c\x75\xca\x5f\x10"
+			"\x17\xe0\x1d\xa3\x66\x50\xe6\x20\x3f\x30"
+			"\xed\x5c\x3d\xa3\xb6\x07\x87\x36\x46\x5e"
+			"\xec\xb4\x00\xee\xaa\xa2\xc8\x76\xe3\x75"
+			"\x64\xd8",
+		.expectedlen = 192,
+		.addtla = NULL,
+		.addtlb = NULL,
+		.addtllen = 0,
+		.pers = NULL,
+		.perslen = 0,
+	},
+};
+
+static const struct drbg_testvec drbg_nopr_sha512_tv_template[] = {
+	{
+		.entropy = (unsigned char *)
+			"\x6b\x50\xa7\xd8\xf8\xa5\x5d\x7a\x3d\xf8"
+			"\xbb\x40\xbc\xc3\xb7\x22\xd8\x70\x8d\xe6"
+			"\x7f\xda\x01\x0b\x03\xc4\xc8\x4d\x72\x09"
+			"\x6f\x8c\x3e\xc6\x49\xcc\x62\x56\xd9\xfa"
+			"\x31\xdb\x7a\x29\x04\xaa\xf0\x25",
+		.entropylen = 48,
+		.expected = (unsigned char *)
+			"\x95\xb7\xf1\x7e\x98\x02\xd3\x57\x73\x92"
+			"\xc6\xa9\xc0\x80\x83\xb6\x7d\xd1\x29\x22"
+			"\x65\xb5\xf4\x2d\x23\x7f\x1c\x55\xbb\x9b"
+			"\x10\xbf\xcf\xd8\x2c\x77\xa3\x78\xb8\x26"
+			"\x6a\x00\x99\x14\x3b\x3c\x2d\x64\x61\x1e"
+			"\xee\xb6\x9a\xcd\xc0\x55\x95\x7c\x13\x9e"
+			"\x8b\x19\x0c\x7a\x06\x95\x5f\x2c\x79\x7c"
+			"\x27\x78\xde\x94\x03\x96\xa5\x01\xf4\x0e"
+			"\x91\x39\x6a\xcf\x8d\x7e\x45\xeb\xdb\xb5"
+			"\x3b\xbf\x8c\x97\x52\x30\xd2\xf0\xff\x91"
+			"\x06\xc7\x61\x19\xae\x49\x8e\x7f\xbc\x03"
+			"\xd9\x0f\x8e\x4c\x51\x62\x7a\xed\x5c\x8d"
+			"\x42\x63\xd5\xd2\xb9\x78\x87\x3a\x0d\xe5"
+			"\x96\xee\x6d\xc7\xf7\xc2\x9e\x37\xee\xe8"
+			"\xb3\x4c\x90\xdd\x1c\xf6\xa9\xdd\xb2\x2b"
+			"\x4c\xbd\x08\x6b\x14\xb3\x5d\xe9\x3d\xa2"
+			"\xd5\xcb\x18\x06\x69\x8c\xbd\x7b\xbb\x67"
+			"\xbf\xe3\xd3\x1f\xd2\xd1\xdb\xd2\xa1\xe0"
+			"\x58\xa3\xeb\x99\xd7\xe5\x1f\x1a\x93\x8e"
+			"\xed\x5e\x1c\x1d\xe2\x3a\x6b\x43\x45\xd3"
+			"\x19\x14\x09\xf9\x2f\x39\xb3\x67\x0d\x8d"
+			"\xbf\xb6\x35\xd8\xe6\xa3\x69\x32\xd8\x10"
+			"\x33\xd1\x44\x8d\x63\xb4\x03\xdd\xf8\x8e"
+			"\x12\x1b\x6e\x81\x9a\xc3\x81\x22\x6c\x13"
+			"\x21\xe4\xb0\x86\x44\xf6\x72\x7c\x36\x8c"
+			"\x5a\x9f\x7a\x4b\x3e\xe2",
+		.expectedlen = 256,
+		.addtla = NULL,
+		.addtlb = NULL,
+		.addtllen = 0,
+		.pers = NULL,
+		.perslen = 0,
+	},
+};
+
 /*
  * ARC4 test vectors from OpenSSL
  */
@@ -34503,7 +36729,94 @@ static const struct cipher_testvec adiantum_xchacha20_aes_tv_template[] = {
 			  "\xe9\xe1\xa0\xfe\xb2\xc3\x80\x77"
 			  "\x20\x05\xe4\x9b\x47\x3b\xb2\xbd",
 		.len	= 4096,
-	}
+	}, {	/* keyLen = 24 */
+		.klen	= 24,
+		.key    = "\x00\x00\x00\x00\x00\x00\x00\x00"
+			  "\x00\x00\x00\x00\x00\x00\x00\x00"
+			  "\x00\x00\x00\x00\x00\x00\x00\x00",
+		.ptext	= "\x1b\x07\x7a\x6a\xf4\xb7\xf9\x82"
+			  "\x29\xde\x78\x6d\x75\x16\xb6\x39"
+			  "\x00\x00\x00\x00\x00\x00\x00\x00"
+			  "\x00\x00\x00\x00\x00\x00\x00\x00"
+			  "\x00\x00\x00\x00\x00\x00\x00\x00"
+			  "\x00\x00\x00\x00\x00\x00\x00\x00"
+			  "\x00\x00\x00\x00\x00\x00\x00\x00"
+			  "\x00\x00\x00\x00\x00\x00\x00\x00"
+			  "\x00\x00\x00\x00\x00\x00\x00\x00"
+			  "\x00\x00\x00\x00\x00\x00\x00\x00"
+			  "\x00\x00\x00\x00\x00\x00\x00\x00"
+			  "\x00\x00\x00\x00\x00\x00\x00\x00"
+			  "\x00\x00\x00\x00\x00\x00\x00\x00"
+			  "\x00\x00\x00\x00\x00\x00\x00\x00"
+			  "\x00\x00\x00\x00\x00\x00\x00\x00"
+			  "\x00\x00\x00\x00",
+		.len	= 124,
+		.ctext	= "\x27\x5c\xfc\x04\x13\xd8\xcc\xb7"
+			  "\x05\x13\xc3\x85\x9b\x1d\x0f\x72"
+			  "\x9f\x88\xa4\x22\xf4\x92\x62\xc8"
+			  "\xb0\xb5\x4c\xfd\x7c\x95\xdd\x5e"
+			  "\x24\x13\x71\x95\xa7\x30\x64\xe6"
+			  "\xed\xbe\x26\x4e\x43\x75\xa9\x7c"
+			  "\xb3\xec\x9a\x54\x46\xc9\x93\x51"
+			  "\xa2\x6b\x06\xcf\x7d\xc9\x30\xe7"
+			  "\x22\xb3\xa2\xf8\x17\x7a\x72\x19"
+			  "\x52\x07\x7c\x5f\x98\xb3\x08\x94"
+			  "\x26\xe1\x96\xe2\x80\xf0\x40\x19"
+			  "\x37\xa6\xd8\xf9\x79\x2d\xdd\x5c"
+			  "\x75\xa0\x06\x0f\x20\x85\x9c\xa0"
+			  "\x14\x6e\xf2\x39\xd3\xc2\x16\xff"
+			  "\xda\xcb\x9e\xde\x36\xa0\x5e\x1f"
+			  "\x7d\x99\x81\x25",
+	}, {	/* keyLen = 32 */
+		.klen	= 32,
+		.key    = "\xc3\xdd\x80\xea\x5c\x74\x4e\x7d"
+			  "\x13\x0a\xa7\x31\x2a\xab\x5d\xc0"
+			  "\xbf\x23\x85\xa7\xfb\xfb\x1a\x4b"
+			  "\x19\x1b\x6e\x08\x55\x78\x7a\x1a",
+		.iv	= "\x07\x43\x84\x47\xe5\x71\x65\xea"
+			  "\x9f\x97\x35\x9a\x84\x1c\x3e\x04",
+		.ptext	= "\xa7\x67\x26\x41\xa2\x9b\x0b\x8d"
+			  "\xeb\x7f\x43\xff\xbd\x7e\x05\x0e"
+			  "\x9f\x8a\x12\xff\x36\xe5\x6b\xe9"
+			  "\x1e\x23\x85\x67\x02\x3a\xfd\x57"
+			  "\x09\x64\xb0\x2f\xb9\xfd\x4a\xd2"
+			  "\x23\xbc\x8d\x67\xe5\x38\x6e\x9e"
+			  "\x83\x70\x42\x52\xcf\x84\x7e\x57"
+			  "\x96\xdf\x3e\xed\x6d\x2e\x49\x79"
+			  "\xe7\x29\x37\x2c\x1d\x88\x7e\xe2"
+			  "\xdd\xe2\x83\xb5\x3c\x02\xc9\x5d"
+			  "\x10\xf2\x96\x4b\xb3\xe4\x80\x41"
+			  "\x90\xcd\xb8\x38\xdf\x28\xb4\xa7"
+			  "\x71\xca\x69\xaf\x96\x7f\x90\x19"
+			  "\xbc\xf6\x37\x50\xc5\x0f\x24\x4f"
+			  "\x17\x1e\x15\x21\xd6\xba\x43\x50"
+			  "\xd8\xce\xac\xae\xa2\x3a\xd8\x14"
+			  "\x32\x49\xc4\x61\x52\xe1\x7c\x26"
+			  "\x14\x6c\x06\xf7\x35\x58\x10\x3c"
+			  "\x7b\x5a\x73\xb8\x3d\xe7\xf0\x36"
+			  "\x2c",
+		.len	= 153,
+		.ctext	= "\x2d\x74\x11\x6f\xab\x68\x44\x71"
+			  "\x87\x2d\xa3\xd3\xd6\x33\x07\x58"
+			  "\x6a\xeb\xac\xc8\xf6\x06\x1c\xd4"
+			  "\x10\x89\xcb\xf3\x50\x21\x0f\x25"
+			  "\x34\x96\x20\x7b\xd1\x07\x81\x90"
+			  "\x30\x80\x65\xe8\xb5\xdf\x90\x8e"
+			  "\xa7\x1d\x69\x61\x40\x80\x36\x95"
+			  "\xf7\xf8\x33\xd1\x5a\xb1\x98\x70"
+			  "\x66\xba\x38\xca\x9b\xf9\xd2\x54"
+			  "\x95\x3c\xed\x59\x3f\xf5\xf5\xc9"
+			  "\x38\xf7\xa6\xfe\xf9\x0c\x78\x49"
+			  "\xd5\x48\x98\x69\xe3\xaa\x85\xbf"
+			  "\x31\xb2\xa3\xb7\x70\xee\x2a\xbf"
+			  "\x92\x4f\xac\x03\xff\x81\x06\xc3"
+			  "\x55\x76\xbe\x8b\x35\x91\x20\x8b"
+			  "\xab\x47\xc1\xe9\x4b\x18\x69\xa4"
+			  "\x10\xce\x3b\x6f\x6b\xc5\x9d\xdf"
+			  "\xf6\xac\x02\x0b\x92\x5e\x84\x61"
+			  "\x7f\x4d\x10\xbf\xdf\xbd\xfa\xec"
+			  "\x7f",
+ 	}
 };
 
 /*
-- 
2.39.4

